{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_community.document_loaders import TextLoader\n",
    "loader = TextLoader(\"speech.txt\")\n",
    "textDocs=loader.load()\n",
    "textDocs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Ignoring wrong pointing object 17 0 (offset 0)\n",
      "Ignoring wrong pointing object 29 0 (offset 0)\n",
      "Ignoring wrong pointing object 32 0 (offset 0)\n",
      "Ignoring wrong pointing object 40 0 (offset 0)\n",
      "Ignoring wrong pointing object 65 0 (offset 0)\n",
      "Ignoring wrong pointing object 73 0 (offset 0)\n",
      "Ignoring wrong pointing object 130 0 (offset 0)\n",
      "Ignoring wrong pointing object 135 0 (offset 0)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[Document(metadata={'source': 'Vae_paper.pdf', 'page': 0}, page_content='Variational Autoencoder for Generation of Antimicrobial PeptidesScott N. Dean and Scott A. Walper*Cite This:ACS Omega2020, 5, 20746−20754Read OnlineACCESSMetrics & MoreArticle Recommendations*sıSupporting Information\\nABSTRACT:Over millennia, natural evolution has allowed for theemergence of countless biomolecules with highly speciﬁc roles withinnatural systems. As seen with peptides and proteins, often evolutionproduces molecules with a similar function but with variable aminoacid composition and structure but diverging from a commonancestor, which can limit sequence diversity. Using antimicrobialpeptides as a model biomolecule, we train a generative deep learningalgorithm on a database of known antimicrobial peptides to generatenovel peptide sequences with antimicrobial activity. Using a variationalautoencoder, we are able to generate a latent space plot that can besurveyed for peptides with known properties and interpolated across apredictive vector between two deﬁned points to identify novelpeptides that show dose-responsive antimicrobial activity. Theseproof-of-concept studies demonstrate the potential for artiﬁcialintelligence-directed methods to generate new antimicrobial peptides and motivate their potential application toward peptideand protein design without the need for exhaustive screening of sequence libraries.1. INTRODUCTIONThe rising specter of broad antibiotic resistance has created acritical need for the rapid development of new antimicrobials.At the current rate, the annual global death toll due toantibiotic resistance is expected to exceed 10 million by 2050at a cost of 100 trillion USD.1One avenue to combat thisproblem is the use of antimicrobial peptides (AMPs). Asintegral components of the innate immune system of humansand other organisms, naturally occurring AMPs have remainedeﬀective antimicrobials despite their ancient origins andwidespread continual contact with pathogens. For this reason,among others, peptide antibiotics including colistin have beendeemed“drugs of last resort”for their ability to kill multidrug-resistant bacteria.2AMPs generally act through mechanisms associated withmembrane disruption, as well as other routes, includingbinding to DNA and essential cytoplasmic proteins, inhibitingtheir normal function.3The relative immutability of bacterialmembranes and other AMP targets makes the development ofresistance to AMPs rare; however, the need for thedevelopment of new AMPs remains essential for combatingmultidrug-resistant bacteria.Many attempts at both generating new AMPs and improvingtheir activity have been made, resulting in some success.4,5These peptides have largely been generated via expertknowledge, low-throughput design methods, including rationaldesign and speciﬁc amino acid substitution, or randomsequence mutation and template-based methods.Various computational tools have been developed forperforming AMP design in a more sophisticated mannerthan those previously used that require expert knowledge,including statistical modeling, quantitative structure−activityrelationship studies, genetic algorithms,6machine learning,7and deep learning.8−10Using these computational methods,key biophysical attributes can be extracted from sequenceinformation, which can be used to provide model creationinformation to predict or further enhance the antimicrobialactivity of AMPs. In general, these tools require large databasesof peptides with known antimicrobial activity. There arecurrently several thousand AMP sequences in various data-bases, including Antimicrobial Peptide Database 3 (APD3)11and A Database of Anti-Microbial (ADAM)12peptides thatcontain AMPs with experimentally determined activity againstGram-negative and Gram-positive bacteria, fungi, HIV, andcancer cells. An extended overview of AMP databases and datamining was recently described by Porto et al.13and newdatabases of various forms are regularly being released.14Since peptides and proteins can be represented as sequentialamino acid residues in the form of a string of characters,several groups have made use of previously developed methodsReceived:January 31, 2020Accepted:May 15, 2020Published:August 10, 2020\\nArticlehttp://pubs.acs.org/journal/acsodf\\nThis article not subject to U.S.Copyright. Published 2020 by theAmerican Chemical Society20746https://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−20754\\nThis is an open access article published under an ACS AuthorChoice License, which permitscopying and redistribution of the article or any adaptations for non-commercial purposes.Downloaded via 106.204.227.228 on August 23, 2023 at 07:04:28 (UTC).See https://pubs.acs.org/sharingguidelines for options on how to legitimately share published articles.'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 1}, page_content='for extracting sequence information in an order-dependentmanner, which have long been used in learning language,including the recursive neural network (RNN) called longshort-term memory (LSTM) networks.15In the fewapplications of LSTMs for peptide design recently reported,the promising ability to generate sequences of active AMPs,with experimental veriﬁcation, has demonstrated the usefulnessof these established deep learning techniques for new AMPgeneration.8,10In another recently developed approach, a number of reportshave demonstrated the utility of generative deep learning,including generative adversarial networks (GANs) and varia-tional autoencoders (VAEs),16for the generation of newobjects from existing data. Among a broad array of attemptedapplications, these have been successfully applied to thegeneration of new images17and new chemical compounds viageneration of simpliﬁed molecular-input line-entry system(SMILES) strings.18,19A noteworthy attribute of VAEs thatdistinguishes it from other generative deep learning techniquesis their ability to create a continuous latent space that can beused to smoothly interpolate between objects. This capabilitywas extensively exhibited by Gómez-Bombarelli et al., wherebyinterpolating between two selected FDA-approved drugs, theygenerated novel chemical structures with a smooth transitionbetween structures, enabling the potential for the optimizationof known drugs, among other applications.18In this study, the latent space generated by a VAE trainedusing an AMP database was utilized for the discovery of newAMPs. The continuous representation of peptides allows forthe generation of novel AMPs in an automated fashion, enablesthe smooth interpolation between AMP sequences, and showsthe potential for optimization of peptide characteristics.Furthermore, we experimentally examine AMPs generated viainterpolation and explore the potential of the method for AMPdesign.2. RESULTS2.1. Model Creation and Generation of ActivePeptides.We implemented and trained a VAE for modelingpeptide sequences with antimicrobial activity. We then usedthe model for generative peptidede novodesign. The generaldesign of the standard VAE architecture is shown inFigure 1.\\nFigure 1.Machine learning for antimicrobial design. (A) Flow diagram of VAE design. During training, both known and scrambled peptides are fedinto the encoder, generating their latent codes (z). The latent codes are then decoded into the peptide sequence by the decoder. The two terms ofthe loss function (KL and reconstruction loss) and model are updated using stochastic gradient descent. (B) Here, we train the VAE using an AMPdataset, which allows for experimental validation of results using common microbiological assays. From starting input AMP sequences, the encodernetwork converts each peptide into a vector in the latent space, which can be viewed as a continuous AMP representation. Provided a point in thelatent space, the decoder network will output a corresponding AMP sequence. The output sequences can then be experimentally tested for activity.\\nFigure 2.Antimicrobial activity assays. (A) Three peptides, P1, P2, and P3, were selected from point locations near short active peptides describedin the literature, NA-CATH, KR-12, and Tet110, generating three peptides sequences designated P1, P2, and P3, respectively, and threecorresponding scrambled control peptides (SP1, SP2, and SP3). (B) Experimental assessment of active (P1−P3) AMPs and scrambled (SP1−SP3)pairs onE. coli,A. baumannii, andS. aureus. Peptide concentrations were 0, 0.64, 3.2, 16, 80, and 400μg/mL. Peptide sequences are listed inTable1. Inhibition of colony formation is indicative of AMP activity.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420747'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 2}, page_content='Five-fold cross-validation on diﬀerent network architectureswas tested for best performance as measured by loss, wherenetworks with 512 or 1024 intermediate dimensionsperformed best, with no observable advantage with largernetworks. Training was stopped after 250 epochs as the lossdecreased at a suﬃciently low rate (less than 0.1 per 10epochs;Figure S1) and results did not noticeably improve withlonger periods of training. Theﬁnal state of the model wassaved and used for sampling novel sequences (Figure 2A).Since the model was trained using two sets of sequencescomposed of“active”and“scrambled”peptides, the implicitstarting assumption was that the sequence order, or“peptidegrammar”,20and characteristics dependent on that sequencewere the components that would be learned by the model.Characteristics related to amino acid frequency, such as netcharge, would not directly contribute to model training asthese are also present in the scrambled peptides.Using the generated latent space, new peptides weredecoded using locations near short active peptides describedin the literature. Latent space positions for the peptides NA-CATH (AP00897; its 11-mer N-terminal variant, ATRA-1, hasbeen extensively studied21), KR-12 (AP00608),22and Tet110(AP02874)23were identiﬁed and used to generate threepeptides sequences designated P1, P2, and P3, respectively.From these sequences, three scrambled control peptides (SP1,SP2, and SP3) were produced (Figure 2A), only altering thesequence order while maintaining amino acid frequency andlength (seeTable 1). Each of these peptides was then assessedfor antimicrobial activity againstEscherichia coli,Acinetobacterbaumannii, andStaphylococcus aureus(Figure 2B).By spot-plating the bacteria following incubation with arange of AMP concentrations, half-maximal eﬀective concen-tration (EC50) values were calculated from triplicate experi-ments. ForE. coli, the EC50values of P1, P2, and P3 weredetermined to be 2.9, 3.1, and 2.9μg/mL, respectively, whileonly SP3 had a determinable EC50of 3.1μg/mL; the others,SP1 and SP2, did not display signiﬁcant killing at theconcentrations of peptide tested (<400μg/mL). The EC50values of P1, P2, and P3 againstA. baumanniiwere found to be3.1, 5.9, and 1.8μg/mL, respectively. Both SP1 and SP3obtained calculable EC50’s of 10.6 and 19μg/mL, respectively,while SP2 did not display antimicrobial activity at theconcentrations used. For the Gram-positiveS. aureus,theEC50values of P1, P2, and P3 were determined to be 0.4, 6.6,and 2.2μg/mL, respectively. Both SP1 and SP3 obtainedcalculable EC50’s of 10.6 and 16.5μg/mL, respectively, whileTable 1. Peptides Used in the Study and Associated Characteristicsname/series increment sequence length molecular weight net charge hydrophobicity hydrophobic momentP1 RKLKKLWRKFR 11 1558.983 6.997−1.782 0.969P2 RRFVKKVRKLVK 12 1557.008 6.997−0.825 0.976P3 FRWLRKWFRR 10 1550.878 4.998−1.430 1.045SP1 KKRRFRWLKLK 11 1558.983 6.997−1.782 0.161SP2 RKLRKKVFVKRV 12 1557.008 6.997−0.825 0.253SP3 WKRLRWRRFF 10 1550.878 4.998−1.430 0.173series 1 0.00 VIREHKYVLLL 11 1382.712 1.090 0.718 0.366series 1 0.33 LPKIKKTVSTR 11 1270.582 3.997−0.682 0.345series 1 0.66 LLKSGRLLMKI 11 1271.671 2.997 0.736 0.637series 1 0.99 KKIKRFLRKIG 11 1386.793 5.997−0.855 1.015series 2 0.00 GLGIIPHRRYGK 12 1366.632 3.088−0.617 0.325series 2 0.33 GIMSLFKGVLKT 12 1293.631 1.997 0.908 0.589series 2 0.66 GLFKIIKNIFSG 12 1336.640 1.997 0.833 0.676series 2 0.99 KLFRIIKRIFKG 12 1518.955 4.997 0.150 1.075\\nFigure 3.Latent space plots. Latent space maps were generated from the APD3 database (plus scrambled sequences). Antimicrobial concept vector(black line) with stepwise sampling shown, extending from the scrambled peptide (A) SP1 (green star) to P1 (red star) and (B) SP2 (green star) toP2 (red star). V1 and V2 are latent variables 1 and 2, respectively. Antimicrobial activity (black arrow) is expected to increase or decreaseproportional to the distance from the scrambled/inactive (SP) or active (P) position. Peptide sequences for each point in the space can be found inTable 1.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420748'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 3}, page_content='SP2 was found to be inactive. EC50values are shown inTableS3. Overall, by mean activity, the scrambled peptides wereassessed as signiﬁcantly less active (p< 0.05; Welch’st-test)than those generated by sampling near active peptides in thelatent space. This result is visible in representative platingimages shown inFigure 2B. This result suggested that themodel learned a degree of sequence-dependent informationconsidering that the scrambled peptides maintained identicalnet charge, hydrophobicity, and other sequence order-independent attributes (listed inTable 1). Of note was thesigniﬁcant diﬀerence in hydrophobic moment between P1, P2,and P3 peptides when compared to SP1, SP2, and SP3: meanhydrophobic moment values of 1.0 and 0.2, respectively. Noneof the other biophysical characteristics examined were found tobe signiﬁcantly diﬀerent between the groups.2.2. Interpolation and Dataset Characterization.Totest the eﬀect of interpolation between active and scrambledpeptides, the overall characteristics of the decoded peptideswere examined to determine the eﬀect of diﬀerent samplingpoints in the latent space. A representation of the latent spacegenerated from 2971 active (obtained from APD311) and 2971corresponding scrambled peptide sequences is shown inFigure3A,B. Here, two series of peptides were generated, betweenSP1 and P1 and between SP2 and P2. Linear interpolation wasperformed, and four peptides were generated along what canbe called an antimicrobial concept vector (black line) instepwise sampling (Figure 3). These four peptide sequenceswere generated at increments of 0.33: 0.00, 0.33, 0.66, and0.99. The sequence at 0.00 was generated close to scrambledpeptide (SP) and the sequence at 0.99 was generated close tothe active peptide (P). For each series, peptides were chosen atrelative distances of 0.33 and 0.66 along the prediction line orantimicrobial concept vector (Table 1). A series was notgenerated for SP3−P3 since both were found to be active.To determine the diﬀerent distributions of characteristics ofthe peptides generated using interpolation, as shown inFigure3, we performed this for every active and scrambled pair in thedataset (2971 pairings, plus the P1−SP1 and P2−SP2 pairs),resulting in a list of 17,838 sequences each with a location label(scrambled, 0.00, 0.33, 0.66, 0.99, and active; seeTable S2).Attributes that are considered to be important for antimicro-bial activity of AMPs, such as hydrophobic moment,24andothers that are likely uncorrelated with activity, such as length,were calculated as a function of step (location of sampling).Although the AMP length was not restricted solely to 12 aminoacids, but was trimmed to≤12, the majority of AMPs were atthat length (96%), with the remainder consisting of shorterpeptides (Figure 4A). Lengths were similar, regardless ofsampling location, with each group containing between 95 and98% peptides of length 12.Like length, the amino acid frequency was similar regardlessof sampling location while not being perfectly uniform (Figure4B). This intergroup similarity was also the case for net charge(Figure 4C), amino acid type (Figure S2A), isoelectric point(pI;Figure S2B), and molecular weight (Figure S2C). Withthe exception of net charge, each of these attributes is generallyconsidered to be largely uncorrelated to antimicrobial activityin AMPs. The net charge within the group was positive (1.24±0.01 (mean±SEM)) ranging between +1.19 and +1.31.While a net positive charge is common among AMPs, between12 and 15% of naturally occurring AMPs are thought to beanionic (net negative charge).25Beyond this, as net charge isan attribute resulting from amino acid frequency and the activeand scrambled group AMPs are identical in this regard, thesameness of the groups in net charge was expected.Peptide hydrophobicity and protein hydrophobicity aredependent on amino acid frequency and structural elements.Within this study and dataset, all groups have a hydrophobicityvalue between 0.08 and 0.2 (overall, 0.13±0.01 (mean±\\nFigure 4.Comparison of peptide characteristics for the entire generated dataset. (A) Length, (B) amino acid frequency, (C) net charge, (D)hydrophobicity, and (E) hydrophobic moment. For (D) and (E), the middle line is displayed at the median. The sequence list for the generateddataset is provided inTable S2.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420749'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 4}, page_content='SEM)) as measured by the Kyte−Doolittle scale.26Interest-ingly, there was an observable trend upward betweenscrambled and active groups (Figure 4D). When examiningthe larger dataset of active and scrambled peptides, at steps0.00, 0.33, 0.66, and 0.99, the hydrophobicity values weredetermined to be 0.08±0.02, 0.11±0.02, 0.14±0.02, and0.19±0.02, respectively. Both scrambled and active groupswere 0.11±0.02. Since overall hydrophobicity is known to bepositively correlated with antimicrobial activity,24this result isnot surprising; however, interestingly, the trend is notcontinued in scrambled and active groups.To estimate the proportion of the secondary structure of thegroups, the GOR IV method was used, which providedpredicted propensities for helix, beta sheet, and random coil.As expected, the average peptide in the active group ispredicted to be primarily helical, while their scrambledcounterparts have a median predicted helicity of∼15%, withthe steps in between increasing in helicity when samplednearer to active (Figure S3). A related, signiﬁcant trend wasobserved in comparing the predicted hydrophobic momentsbetween the sampling step locations. The hydrophobicmoments were calculated using a default angle of 100°,which assumes a helical secondary structure. Hydrophobicmoments for each group were found to be 0.26±0.002, 0.28±0.002, 0.30±0.003, 0.32±0.003, 0.35±0.003, and 0.38±0.003 for scrambled, 0.00, 0.33, 0.66, 0.99, and active,respectively. This means that the spatial orientations ofhydrophobic and polar amino acids diﬀer between the groupsalong the interpolation path, leading to signiﬁcantly increasedhydrophobic moments (p< 0.01, two-sided Welch’st-test)compared to those of the scrambled peptide set (Figure 4E).Also, importantly, the mean value of each group is signiﬁcantlydiﬀerent from each adjacent set, suggesting a gradual shift inspatial orientations along the route from scrambled to active.Next, to quantify the potential antimicrobial activity of thegenerated peptides, we made use of the CAMPR3 AMPprediction tool. We decided to use the CAMPR3 classiﬁer as ithas been empirically determined to be superior to otherclassiﬁers in a recent benchmark report,27best reﬂecting realantimicrobial activity as determined in experiments. Theprobability of antimicrobial activity (P(AMP)) of the AMPswas obtained and plotted inFigure 5A. In a trend similar to thehydrophobic moment described above, theP(AMP) increasedas a function of the sampling step locations between scrambledand active. TheP(AMP) for each group was, as output byCAMPR3, 0.40±0.006, 0.41±0.006, 0.41±0.006, 0.44±0.006, 0.49±0.007, and 0.50±0.007 for scrambled, 0.00,0.33, 0.66, 0.99, and active, respectively. Meanwhile, the overallP(AMP) was 0.44±0.002 (mean±SEM), suggesting that, onaverage, the peptides generated in this study are predicted tobe inactive (where aP(AMP) of≥0.5 is classiﬁed as active).However, several of these are clearly active, as seen inFigure5B and discussed below. Critically, Welch’st-test rejected thehypothesis that the scrambled and active datasets have equalmeans of activity predictions (p< 0.05), as well as 0.66 and0.99 datasets when compared to the scrambled dataset (p<0.05). Interestingly, at the 0.33 point, the peptides were not ofhigherP(AMP).The relatively lowP(AMP) prediction for those sequencesin the active peptide group may result from the following: (1)the sequences in the training set for CAMPR3 are substantiallylonger on average than those used in this study (≤12), (2)short peptides may not contain suﬃcient information for mostmodels to properly assess likelihood of activity, and (3) it ispossible that the truncation of peptides in the APD3-baseddataset did impact real activity. Overall, according to theCAMPR3 predictions, the VAE model can be used tosmoothly interpolate between AMPs predicted to be active(or borderline active) and those that are predicted to beinactive.2.3. Evaluation of Antimicrobial Activity.As describedabove (Section 2.2), following their testing, the peptides P1,P2, SP1, and SP2 were used as markers to test the capability ofinterpolating between the two groups, scrambled and active.Two series of interpolating peptides were generated at fourincrements between each of the scrambled−active pairs of P1−SP1 and P2−SP2 at increments of 0.00, 0.33, 0.66, and 0.99, asshown inFigure 3.T h er e s u l t i n gp e p t i d e sw e r en e x texperimentally investigated by assessing their antimicrobialactivities onE. coli,A. baumannii, andS. aureus(Figure 5B). Byspot-plating the bacteria following incubation with a range ofAMP concentrations, EC50’s were calculated from triplicateexperiments. EC50’s for both series 1 and 2 peptides againsteach of the three species tested are found inTable S4andcorresponding EC50curves are shown inFigure S4.ForE. coli, the EC50values of peptides 0.66 and 0.99 forseries 1 were found to be 308 and 11μg/mL, respectively,while neither preceding peptides in the series displayed\\nFigure 5.(A) Active AMP prediction using CAMPR3. (B) Peptide series 1 and 2 tested againstE. coli,A. baumannii, andS. aureus.Peptideconcentrations were 0, 0.64, 3.2, 16, 80, and 400μg/mL. Row labels are interpolation step for each peptide series. The absence or inhibition ofcolony formation is indicative of AMP activity.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420750'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='activity. Similarly, for series 2, peptides 0.66 and 0.99 werefound to have EC50’s of 84 and 0.2μg/mL, respectively, whileboth peptides 0.00 and 0.33 were inactive at the concentrationstested. When tested againstA. baumannii, the EC50values ofpeptides 0.66 and 0.99 for series 1 were found to be 400 and36μg/mL, respectively, while neither of the preceding peptides0.00 or 0.33 display activity. For series 2, the EC50’s wereobserved to be 65, 13, and 0.8μg/mL for 0.33, 0.66, and 0.99,respectively; peptide 0.00 was inactive. ForS. aureus,the EC50values of peptides 0.33, 0.66, and 0.99 for series 1 weredetermined as 69, 11, and 0.4μg/mL, respectively, whilepeptide 0.00 did not show activity. For series 2, all of thepeptides were found inactive at the concentrations examined.Overall, these results appear to conﬁrm theP(AMP)predictions by CAMPR3 inFigure 5A, suggesting that thepeptides sampled from regions closer to the active space inlatent space are more active than those decoded in closeproximity to scrambled. This trend is further supported bycomparing the overall mean activities, which show that 0.99group AMPs have a signiﬁcantly lower EC50than those from0.66 (p< 0.05; Welch’st-test). Although the EC50values formany of those in group 0.33 and all of those in group 0.00 arenot available due to insuﬃcient activity for calculation,Figure5B suggests the overall trend of increasing activity along thepath between scrambled and active groups.3. DISCUSSIONThis study demonstrates the use of a VAE forde novopeptidesequence design using AMPs as an example. The VAEdescribed has the potential for sequence design automationprovided that a large database of AMP sequences is available, arequirement similar to other computational techniques. Theincreasing number of both AMP databases and entries perdatabase makes AMPs a viable proof of concept for assessingthe ability of VAEs to generate new peptides. This approach isdistinct from many previously reported design techniques, suchas random mutagenesis, template-based peptide designstrategies, and other random sequence generation methods,where amino acids are drawn from a predeﬁned distributionand then concatenated.5The encoder and decoder of the VAErely on LSTM RNNs that have high-dimensional sequencerepresentations, meaning that it does not simply reproducesequence templates but learns important features present in thetraining dataset.The results achieved suggest that the peptide VAE modelwas capable of generating an internal representation of AMPsfrom the APD3 database, in that the model did not simplyreproduce the training data, supported by the absence of anypreviously known sequences in those experimentally examined.However, when sampled from similar locations in the latentspace, the decoded peptides do have statistically similardistributions to those found in the training data (either activeor scrambled). Critically, the features generally reported to beimportant for the antimicrobial activity of AMPs, whether forbacterial membrane targeting or otherwise, such as highhydrophobic moment in alpha helical peptides,24are retainedin the peptides decoded near active peptides, while propertiesless correlated with antimicrobial activity, such as length,molecular weight, and pI, are roughly constant regardless oflocation in the latent space. In fact, of all of the peptideattributes examined in the study, the main diﬀerence betweenthe active model-generated sequences and scrambled randomsequences is their amphipathicity, which is illustrated by ahigher hydrophobic moment for the model-generated setsclose to active peptides in the latent space, consistent with theresults reported by Müller et al. utilizing a LSTM RNNmodel.8Since the hydrophobic moment translates back to a regularpattern of charged and hydrophobic residues at speciﬁclocations in the sequence, this demonstrates that this hasbeen learned by the model. It is very likely that the increasinghydrophobic moment yielded higher probabilities in theCAMPR3 AMP prediction model,28suggesting that simplyalternating groups of positively ionizable and hydrophobicamino acids will score highly by the CAMPR3 predictor, ahypothesis that has been manipulated by Müller et al.,8highlighting the importance of experimentally verifying theantimicrobial activity of generated AMPs. This, along withpreserving the simplicity of the method presented, is a possibleadvantage of avoiding pairing generative models withpredictive models, which has been done elsewhere:29we canavoid any possibility of forcing the generative model to“learn”to tailor its output to what scores highly in a predictive model.To verify antimicrobial activity, the interpolation featureprovided by VAEs was exploited. By interpolating between twoknown peptides, an active peptide found in APD3 and anassumed inactive scrambled partner, increasing gradations ofantimicrobial activity can be sampled at regular steps along theline (Figure 2). This study demonstrated that two series ofpeptides, between P1−SP1 and P2−SP2, sampled at four stepseach could yield an increasing spectrum of activity betweentwo points. The results indicate that not only did the VAEactually learn to recognize the grammar of amphipathicity inpeptides but also it may have learned the underlying grammarof what features contribute to antimicrobial activity. Thecriticalﬁnding is that the VAE learned the grammar ofamphipathicity and antimicrobial activity in peptides, whichcan then be tuned in a controllable manner.As observed in these studies, the intermediates of P3−SP3displayed high antimicrobial activity as both the predicted andscrambled peptides. With the P1, P2, SP1, and SP2 constructs,the reduced activity of the scrambled peptides can likely beattributed to a change in the hydrophobic moment and,therefore, the ability for the peptide to readily insert into thebacterial membrane. For the P3−SP3 pair, however, thesigniﬁcant change in hydrophobic moment does not appear toaﬀect the antimicrobial activity. The P3−SP3 peptides arecomposed of multiple tryptophan and arginine residues. In apublication by Wessolowski et al., a similar phenotype wasobserved with short peptides of similar composition,suggesting antimicrobial activity through noncanonical mech-anisms.30While their consistent activity as native andscrambled peptides is interesting, neighboring peptides werenot examined in these studies as similar activity was expected.This observation, however, illustrates the potential need forpairing the VAE with aﬁltering algorithm to remove certainclasses of AMPs depending on the application.As conﬁrmation of the active AMP prediction model,28thetwo interpolated series of AMPs were tested againstE. coli,A.baumannii, andS. aureusand, overall, yielded the same trendidentiﬁed by the AMP prediction, where peptides sampleddisplayed more activity as a function of their proximity toknown active peptides. Interestingly, the peptide series mayshow a Gram bias, in that the series 2 peptides did not displayany measurable activity againstS. aureusbut were active againstGram-negative bacteriaE. coliandA. baumannii.MoreACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420751'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='investigation is required to determine if this result waspredictable prior to synthesis. Although, for simplicity, nopredictive model was paired with VAE in the current study,various other groups using generative models for chemical orpeptide generation have output shuttled through predictivemodels prior to use.18Such a pairing could be developed forfuture applications to potentially wield some control overGram bias in antimicrobial activity and other attributes ofpeptides prior to experiment assessment.4. CONCLUSIONSThe results of the study demonstrate the use of a VAE for theautomated generation of novel AMPs. The model was trainedon thousands of known and scrambled AMP sequences fromAPD3 and created a readily usable tool, the form of acontinuous representation of peptides. The latent spaceallowed us to generate novel AMPs in a controllable fashion,which we experimentally veriﬁed to be active. Despite theshortcomings of this proof-of-concept study, including thefocus on truncated peptides, this favorable result will drivefuture work in the application of VAEs toward various peptideand protein-designing projects, including full-length AMPs.This work suggests that the VAE could operate with greaterutility when paired with a predictive model, or othercharacteristic classiﬁer, in a manner described by Gómez-Bombarelli et al. for molecule generation to better predictAMP activity and for optimization purposes. Beyond this, it isunclear whether the VAE technique is speciﬁcally well suitedfor peptide and protein applications or could other recentlydeveloped and improved generative deep learning techniquessuch as GANs result in superior performance. This idea, use ofa conditional VAE or otherwise pairing with a characteristicclassiﬁer, and assessment of VAE-generated AMPs on otherbacteria will be the subject of future research.5. EXPERIMENTAL SECTION5.1. Dataset and Preprocessing.AMP sequences andother information, including activity (Gram-negative bacteria,Gram-positive bacteria, cancer cells, etc.), were downloadedfrom APD3,11yielding 2971 peptide sequences (accessed inmid-2018). Despite theﬂaws of APD3, such as highhomogeneity and incomplete annotation of post-translationallymodiﬁed sequences, it remains to be one of largest AMPdatabases available. Since synthesis of long peptides can beprohibitively expensive, and to reduce the computational loaddownstream, the sequences were truncated such that allpeptides were≤12 amino acids in length, retaining the N-terminal sequence. The length of≤12 amino acids was chosenbased on several reports that have shown that short (10−12amino acids), truncated AMPs are capable of maintaining highlevels of antimicrobial activity.4,31,32The list of trimmedpeptide sequences was then duplicated and randomlyscrambled with the stringi R package21function stri_rand_-shuﬄe to create a scrambled peptide pair for each peptidefound in APD3, resulting in a list of 5942 sequences. Duplicateactive peptides were retained as all scrambled pairings wereunique. The sequences were then tokenized (each residueseparated; seeTable S1) and represented by a one-hotencoding scheme using binary vectors with length equal to thesize of the amino acid vocabulary:“<end>”,“a”,“c”,“d”,“e”,“f’”,“g”,“h”,“i”,“k”,“l”,“m”,“n”,“p”,“q”,“r”,“s”,“t”,“v”,“w”,“x”, and“y”, where“x”is an unknown amino acid present insome of the APD3 sequences. This resulted in a 3D datamatrix of dimensions 5942, 24, and 12 for the number ofsequences, length of the vocabulary, and feature vector length,respectively.5.2. Training and Sequence Generation.The archi-tecture of the VAE was implemented as described by Bowmanet al.,33including use of character dropout, which replaces arandom fraction of characters (listed inSection 5.1) with an<unk> character and utilization of Kullback−Leibler (KL)annealing, comprising an additional variable weight added tothe cost function KL term while training. The loss function, asdescribed by Bowman et al., was composed of reconstructionloss and KL loss to penalize poor reconstruction of the data bythe decoder and encoder output representations of z (latentspace variables) that are diﬀerent from a standard normaldistribution, respectively (Figure 1A). Training stoppagecriteria were met when loss values did not decrease for 10iterations through the entire dataset (epochs). The prepro-cessed data were encoded into vectors using LSTMs. Theencoder LSTM was paired with a decoder LSTM to dosequence-to-sequence learning. The decoder results wereconverted from binary one-hot encoded vectors to peptidesequences (Figure 1B). The VAE was trained using the Keras34library with a TensorFlow35backend via the Adam optimizer.The number of neurons for the LSTM layers found in bothencoder and decoder was both set to 1024. All models weretrained on an Ubuntu workstation with a Nvidia GeforceGTX1070 GPU. The LSTM RNNs used in the decoder (andencoder) are stochastic, meaning that decoding from the samepoint in the latent space may result in a diﬀerent peptide beinggenerated and is dependent on the random seed set prior torunning. Sequence sampling was performed using linearinterpolation between active peptides and their scrambledpairs, which is expected to be inactive or less functional as anAMP against target bacteria. For each pairing, four points weregenerated at equally spaced increments of 0.33: 0.00, 0.33,0.66, and 0.99, where the scrambled pair is encoded at point 0and the active pair is encoded at 1 such that points 0.00 and0.99 are adjacent to AMPs in the dataset, while points decodedat 0.33 and 0.66 are intermediates. Here, it was theorized thatantimicrobial activity would correspond to the latent spaceposition or distance from either the active (position 1.0) orinactive/scrambled peptide (position 0.0).5.3. Examination of Sequence Characteristics.Thecharacteristics of sequences generated, including peptidelength, amino acid composition, net charge, hydrophobicity,and hydrophobic moment, were assessed using the Peptides Rpackage36and Python library modlAMP.37Prediction of thethree-state protein secondary structure (helix, beta, and coil)was carried out using the Garnier−Osguthorpe−Robson(GOR) IV method38implemented in the Decipher Rpackage.39The probability of AMP activity (P(AMP)) waspredicted using CAMPR327by uploading sequences to itsserver. Exported results from CAMPR3, as well as character-istic values output from the Peptides package and modlAMP,were visualized using ggplot2 in the tidyverse R package.405.4. Antimicrobial Assays.Peptides synthesized for use inthis study are listed inTable 1. Peptides were synthesized byGenscript, Inc. (Piscataway, NJ), and each was conﬁrmed tohave greater than 90% purity. Lyophilized peptides weresolubilized in water, aliquoted, and stored at−20°C.Overnight cultures ofE. coliBL21,A. baumanniiATCC17978, andS. aureusATCC 12600 were grown in LysogenyACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420752'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 7}, page_content='Broth (LB). EC50assays were performed as previouslydescribed.41Overnight cultures of bacteria were enumeratedby a standard curve using optical density vs CFU/mL. Cultureswere resuspended in 10 mM sodium phosphate buﬀer (pH 7)to aﬁnal concentration of∼2×106CFU/mL. Fordetermination of EC50values, peptides were assessed at arange of concentrations (0.64−400μg/mL) with bacteria for 1h at 37°C. Following incubation, bacteria were spotted forcolony counting as previously described;42full spread platingwas performed for conﬁrmation of colony counting.■ASSOCIATED CONTENT*sıSupporting InformationThe Supporting Information is available free of charge athttps://pubs.acs.org/doi/10.1021/acsomega.0c00442.EC50values for P1, P2, and P3 and their scrambled pairs(Table S3); EC50values corresponding to the survivalcurves found in Figure S3 (Table S4); representative losscurve from the training of the VAE (Figure S1);comparison of peptide characteristics (Figure S2);secondary structure prediction of generated peptidesacross each group (Figure S3); and EC50curves for thepeptide series examined (Figure S4) (PDF)Input dataset for training (Table S1) (XLSX)Generated peptides (Table S2) (XLSX)■AUTHOR INFORMATIONCorresponding AuthorScott A. Walper−Center for Bio/Molecular Science&Engineering (Code 6900), U.S. Naval Research Laboratory,Washington, D.C. 20375, United States;orcid.org/0000-0002-9436-3456; Email:scott.walper@nrl.navy.milAuthorScott N. Dean−National Research Council Associate,Washington, D.C. 20001, United StatesComplete contact information is available at:https://pubs.acs.org/10.1021/acsomega.0c00442Author ContributionsThe manuscript was written through contributions of allauthors. All authors have given approval to theﬁnal version ofthe manuscript.NotesThe authors declare no competingﬁnancial interest.■ACKNOWLEDGMENTSS.N.D. acknowledges the funding support through theNational Research Council Research Associateship Programand additional funding through the Jerome and Isabella KarleDistinguished Scholar Fellowship provided by the NavalResearch Laboratory. S.A.W. was supported by Core fundsprovided to the Naval Research Laboratory through Oﬃce ofNaval Research (WU#MA041−06−41).■ABBREVIATIONSAMPs, antimicrobial peptides; VAE, variational autoencoder;RNN, recursive neural network; LSTM, long short-termmemory; EC50, half-maximal eﬀective concentration; SMILES,simpliﬁed molecular-input line-entry system■REFERENCES(1) O’Neill, J. Antimicrobial Resistance: Tackling a crisis for thehealth and wealth of nations.Rev. Antimicrob. Resist.2014, 1.(2) Lewies, A.; Du Plessis, L. H.; Wentzel, J. F. AntimicrobialPeptides: the Achilles’Heel of Antibiotic Resistance?ProbioticsAntimicrob. Proteins2019, 370−381.(3) Chung, E. M. C.; Dean, S. N.; Propst, C. N.; Bishop, B. M.; vanHoek, M. L. Komodo dragon-inspired synthetic peptide DRGN-1promotes wound-healing of a mixed-biofilm infected wound.npjBiofilms Microbiomes2017,3, 9.(4) Dean, S. N.; Bishop, B. M.; van Hoek, M. L. Natural andsynthetic cathelicidin peptides with anti-microbial and anti-biofilmactivity against Staphylococcus aureus.BMC Microbiol.2011,11, 114.(5) Torres, M. D. T.; Sothiselvam, S.; Lu, T. K.; de la Fuente-Nunez,C. Peptide Design Principles for Antimicrobial Applications.J. Mol.Biol.2019, 3547.(6) Cardoso, M. H.; Oshiro, K. G. N.; Rezende, S. B.; Cândido, E.S.; Franco, O. L. The Structure/Function Relationship in Anti-microbial Peptides: What Can we Obtain From Structural Data?Adv.Protein Chem. Struct. Biol.2018,112, 359−384.(7) Lima, A. N.; Philot, E. A.; Trossini, G. H. G.; Scott, L. P. B.;Maltarollo, V. G.; Honorio, K. M. Use of machine learning approachesfor novel drug discovery.Expert Opin. Drug Discovery2016,11, 225−239.(8) Müller, A. T.; Hiss, J. A.; Schneider, G. Recurrent NeuralNetwork Model for Constructive Peptide Design.J. Chem. Inf. Model.2018,58, 472−479.(9) Nagarajan, D.; Nagarajan, T.; Roy, N.; Kulkarni, O.;Ravichandran, S.; Mishra, M.; Chakravortty, D.; Chandra, N.Computational antimicrobial peptide design and evaluation againstmultidrug-resistant clinical isolates of bacteria.J. Biol. Chem.2018,293, 3492−3509.(10) Schneider, P.; Müller, A. T.; Gabernet, G.; Button, A. L.;Posselt, G.; Wessler, S.; Hiss, J. A.; Schneider, G. Hybrid NetworkModel for″Deep Learning″of Chemical Data: Application toAntimicrobial Peptides.Mol. Inf.2017,36, 1600011.(11) Wang, G.; Li, X.; Wang, Z. APD3: the antimicrobial peptidedatabase as a tool for research and education.Nucleic Acids Res.2016,44, D1087−D1093.(12) Lee, H.-T.; Lee, C.-C.; Yang, J.-R.; Lai, J. Z. C.; Chang, K. Y. Alarge-scale structural classification of antimicrobial peptides.BioMedRes. Int.2015,2015,1−6.(13) Porto, W. F.; Pires, A. S.; Franco, O. L. Computational tools forexploring sequence databases as a resource for antimicrobial peptides.Biotechnol. Adv.2017,35, 337−349.(14) Nagarajan, D.; Nagarajan, T.; Nanajkar, N.; Chandra, N. Auniform in vitro efficacy dataset to guide antimicrobial peptide design.Data2019,4, 27.(15) Gers, F. A.; Schmidhuber, E. LSTM recurrent networks learnsimple context-free and context-sensitive languages.IEEE Trans.Neural Networks2001,12, 1333−1340.(16) Kingma, D. P.; Welling, M. Auto-encoding variational bayes.2013, arXiv:1312.6114. arXiv.org e-Print archive.https://arxiv.org/abs/1312.6114.(17) White, T. Sampling generative networks.2016,arXiv:1609.04468. arXiv.org e-Print archive.https://arxiv.org/abs/1609.04468.(18) Gömez-Bombarelli, R.; Wei, J. N.; Duvenaud, D.; Hernández-Lobato, J. M.; Sánchez-Lengeling, B.; Sheberla, D.; Aguilera-Iparraguirre, J.; Hirzel, T. D.; Adams, R. P.; Aspuru-Guzik, A.Automatic chemical design using a data-driven continuous repre-sentation of molecules.ACS Cent. Sci.2018,4, 268−276.(19) Kusner, M. J.; Paige, B.; Hernández-Lobato, J. M. Grammarvariational autoencoder.InProceedings of the 34th InternationalConference on Machine Learning-Volume 70, JMLR.org: 2017; pp1945−1954.(20) Loose, C.; Jensen, K.; Rigoutsos, I.; Stephanopoulos, G. Alinguistic model for the rational design of antimicrobial peptides.Nature2006,443, 867.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420753'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 8}, page_content='(21) Dean, S. N.; Bishop, B. M.; van Hoek, M. L. Susceptibility ofPseudomonas aeruginosa Biofilm to Alpha-Helical Peptides: D-enantiomer of LL-37.Front. Microbiol.2011,2, 128.(22) Wang, G. Structures of human host defense cathelicidin LL-37and its smallest antimicrobial peptide KR-12 in lipid micelles.J. Biol.Chem.2008,283, 32637−32643.(23) Hilpert, K.; Elliott, M.; Jenssen, H.; Kindrachuk, J.; Fjell, C. D.;Körner, J.; Winkler, D. F. H.; Weaver, L. L.; Henklein, P.; Ulrich, A.S.; Chiang, S. H. Y.; Farmer, S. W.; Pante, N.; Volkmer, R.; Hancock,R. E. W. Screening and characterization of surface-tethered cationicpeptides for antimicrobial activity.Chem. Biol.2009,16, 58−69.(24) Pathak, N.; Salas-Auvert, R.; Ruche, G.; Janna, M. h.;McCarthy, D.; Harrison, R. G. Comparison of the effects ofhydrophobicity, amphiphilicity, andα-helicity on the activities ofantimicrobial peptides.Proteins: Struct., Funct., Bioinf.1995,22, 182−186.(25) Harris, F.; Dennison, S. R.; Phoenix, D. A. Anionicantimicrobial peptides from eukaryotic organisms.Curr. ProteinPept. Sci.2009,10, 585−606.(26) Kyte, J.; Doolittle, R. F. A simple method for displaying thehydropathic character of a protein.J. Mol. Biol.1982,157, 105−132.(27) Gabere, M. N.; Noble, W. S. Empirical comparison of web-based antimicrobial peptide prediction tools.Bioinformatics2017,33,1921−1929.(28) Waghu, F. H.; Barai, R. S.; Gurung, P.; Idicula-Thomas, S.CAMPR3: a database on sequences, structures and signatures ofantimicrobial peptides.Nucleic Acids Res.2016,44, D1094−D1097.(29) Xu, Y.; Lin, K.; Wang, S.; Wang, L.; Cai, C.; Song, C.; Lai, L.;Pei, J. Deep learning for molecular generation.Future Med. Chem.2019, 567.(30) Wessolowski, A.; Bienert, M.; Dathe, M. Antimicrobial activityof arginine- and tryptophan-rich hexapeptides: the effects of aromaticclusters, D-amino acid substitution and cyclization.J. Pept. Res.2004,64, 159−169.(31) Bruni, N.; Capucchio, M.; Biasibetti, E.; Pessione, E.;Cirrincione, S.; Giraudo, L.; Corona, A.; Dosio, F. Antimicrobialactivity of lactoferrin-related peptides and applications in human andveterinary medicine.Molecules2016,21, 752.(32) Kasetty, G.; Papareddy, P.; Kalle, M.; Rydengård, V.; Mörgelin,M.; Albiger, B.; Malmsten, M.; Schmidtchen, A. Structure-activitystudies and therapeutic potential of host defense peptides of humanthrombin.Antimicrob. Agents Chemother.2011,55, 2880−2890.(33) Bowman, S. R.; Vilnis, L.; Vinyals, O.; Dai, A. M.; Jozefowicz,R.; Bengio, S. Generating sentences from a continuous space.2015,arXiv:1511.06349. arXiv.org e-Print archive.https://arxiv.org/abs/1511.06349.(34) Chollet, F.Keras.https://github.com/fchollet/keras.(35) Abadi, M.; Barham, P.; Chen, J.; Chen, Z.; Davis, A.; Dean, J.;Devin, M.; Ghemawat, S.; Irving, G.; Isard, M.Tensorﬂow: A systemfor large-scale machine learning.In12th {USENIX} Symposium onOperating Systems Design and Implementation ({OSDI} 16); USENIXAssociation: 2016; pp 265−283.(36) Osorio, D.; Rondón-Villarrea, P.; Torres, R. Peptides: a packagefor data mining of antimicrobial peptides.R Journal2015,7,4−14.(37) Müller, A. T.; Gabernet, G.; Hiss, J. A.; Schneider, G.modlAMP: Python for antimicrobial peptides.Bioinformatics2017,33, 2753−2755.(38) Garnier, J.; Gibrat, J.-F.; Robson, B. GOR method forpredicting protein secondary structure from amino acid sequence.InMethods in enzymology, Elsevier: 1996; Vol.266, pp 540−553,DOI: 10.1016/S0076-6879(96)66034-0.(39) Wright, E. S. DECIPHER: harnessing local sequence context toimprove protein multiple sequence alignment.BMC Bioinf.2015,16,322.(40) Wickham, H.Tidyverse: Easily install and load’tidyverse’packages.R package version; R Core Team: 2017,1(1).(41) Chung, M. C.; Dean, S. N.; van Hoek, M. L. Acyl carrierprotein is a bacterial cytoplasmic target of cationic antimicrobialpeptide LL-37.Biochem. J.2015,470, 243−253.(42) Sieuwerts, S.; de Bok, F. A. M.; Mols, E.; de vos, W. M.; vanHylckama Vlieg, J. E. T. A simple and fast method for determiningcolony forming units.Lett. Appl. Microbiol.2008,47, 275−278.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420754')]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain_community.document_loaders import PyPDFLoader\n",
    "loader = PyPDFLoader(\"Vae_paper.pdf\")\n",
    "docs=loader.load()\n",
    "docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(metadata={'source': 'Vae_paper.pdf', 'page': 0}, page_content='Variational Autoencoder for Generation of Antimicrobial PeptidesScott N. Dean and Scott A. Walper*Cite This:ACS Omega2020, 5, 20746−20754Read OnlineACCESSMetrics & MoreArticle Recommendations*sıSupporting Information'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 0}, page_content='ABSTRACT:Over millennia, natural evolution has allowed for theemergence of countless biomolecules with highly speciﬁc roles withinnatural systems. As seen with peptides and proteins, often evolutionproduces molecules with a similar function but with variable aminoacid composition and structure but diverging from a commonancestor, which can limit sequence diversity. Using antimicrobialpeptides as a model biomolecule, we train a generative deep learningalgorithm on a database of known antimicrobial peptides to generatenovel peptide sequences with antimicrobial activity. Using a variationalautoencoder, we are able to generate a latent space plot that can besurveyed for peptides with known properties and interpolated across apredictive vector between two deﬁned points to identify novelpeptides that show dose-responsive antimicrobial activity. Theseproof-of-concept studies demonstrate the potential for artiﬁcialintelligence-directed methods to generate new antimicrobial peptides and'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 0}, page_content='that show dose-responsive antimicrobial activity. Theseproof-of-concept studies demonstrate the potential for artiﬁcialintelligence-directed methods to generate new antimicrobial peptides and motivate their potential application toward peptideand protein design without the need for exhaustive screening of sequence libraries.1. INTRODUCTIONThe rising specter of broad antibiotic resistance has created acritical need for the rapid development of new antimicrobials.At the current rate, the annual global death toll due toantibiotic resistance is expected to exceed 10 million by 2050at a cost of 100 trillion USD.1One avenue to combat thisproblem is the use of antimicrobial peptides (AMPs). Asintegral components of the innate immune system of humansand other organisms, naturally occurring AMPs have remainedeﬀective antimicrobials despite their ancient origins andwidespread continual contact with pathogens. For this reason,among others, peptide antibiotics including colistin have'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 0}, page_content='AMPs have remainedeﬀective antimicrobials despite their ancient origins andwidespread continual contact with pathogens. For this reason,among others, peptide antibiotics including colistin have beendeemed“drugs of last resort”for their ability to kill multidrug-resistant bacteria.2AMPs generally act through mechanisms associated withmembrane disruption, as well as other routes, includingbinding to DNA and essential cytoplasmic proteins, inhibitingtheir normal function.3The relative immutability of bacterialmembranes and other AMP targets makes the development ofresistance to AMPs rare; however, the need for thedevelopment of new AMPs remains essential for combatingmultidrug-resistant bacteria.Many attempts at both generating new AMPs and improvingtheir activity have been made, resulting in some success.4,5These peptides have largely been generated via expertknowledge, low-throughput design methods, including rationaldesign and speciﬁc amino acid substitution, or randomsequence'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 0}, page_content='in some success.4,5These peptides have largely been generated via expertknowledge, low-throughput design methods, including rationaldesign and speciﬁc amino acid substitution, or randomsequence mutation and template-based methods.Various computational tools have been developed forperforming AMP design in a more sophisticated mannerthan those previously used that require expert knowledge,including statistical modeling, quantitative structure−activityrelationship studies, genetic algorithms,6machine learning,7and deep learning.8−10Using these computational methods,key biophysical attributes can be extracted from sequenceinformation, which can be used to provide model creationinformation to predict or further enhance the antimicrobialactivity of AMPs. In general, these tools require large databasesof peptides with known antimicrobial activity. There arecurrently several thousand AMP sequences in various data-bases, including Antimicrobial Peptide Database 3 (APD3)11and A Database of'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 0}, page_content='databasesof peptides with known antimicrobial activity. There arecurrently several thousand AMP sequences in various data-bases, including Antimicrobial Peptide Database 3 (APD3)11and A Database of Anti-Microbial (ADAM)12peptides thatcontain AMPs with experimentally determined activity againstGram-negative and Gram-positive bacteria, fungi, HIV, andcancer cells. An extended overview of AMP databases and datamining was recently described by Porto et al.13and newdatabases of various forms are regularly being released.14Since peptides and proteins can be represented as sequentialamino acid residues in the form of a string of characters,several groups have made use of previously developed methodsReceived:January 31, 2020Accepted:May 15, 2020Published:August 10, 2020'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 0}, page_content='Articlehttp://pubs.acs.org/journal/acsodf\\nThis article not subject to U.S.Copyright. Published 2020 by theAmerican Chemical Society20746https://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−20754\\nThis is an open access article published under an ACS AuthorChoice License, which permitscopying and redistribution of the article or any adaptations for non-commercial purposes.Downloaded via 106.204.227.228 on August 23, 2023 at 07:04:28 (UTC).See https://pubs.acs.org/sharingguidelines for options on how to legitimately share published articles.'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 1}, page_content='for extracting sequence information in an order-dependentmanner, which have long been used in learning language,including the recursive neural network (RNN) called longshort-term memory (LSTM) networks.15In the fewapplications of LSTMs for peptide design recently reported,the promising ability to generate sequences of active AMPs,with experimental veriﬁcation, has demonstrated the usefulnessof these established deep learning techniques for new AMPgeneration.8,10In another recently developed approach, a number of reportshave demonstrated the utility of generative deep learning,including generative adversarial networks (GANs) and varia-tional autoencoders (VAEs),16for the generation of newobjects from existing data. Among a broad array of attemptedapplications, these have been successfully applied to thegeneration of new images17and new chemical compounds viageneration of simpliﬁed molecular-input line-entry system(SMILES) strings.18,19A noteworthy attribute of VAEs thatdistinguishes it'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 1}, page_content='to thegeneration of new images17and new chemical compounds viageneration of simpliﬁed molecular-input line-entry system(SMILES) strings.18,19A noteworthy attribute of VAEs thatdistinguishes it from other generative deep learning techniquesis their ability to create a continuous latent space that can beused to smoothly interpolate between objects. This capabilitywas extensively exhibited by Gómez-Bombarelli et al., wherebyinterpolating between two selected FDA-approved drugs, theygenerated novel chemical structures with a smooth transitionbetween structures, enabling the potential for the optimizationof known drugs, among other applications.18In this study, the latent space generated by a VAE trainedusing an AMP database was utilized for the discovery of newAMPs. The continuous representation of peptides allows forthe generation of novel AMPs in an automated fashion, enablesthe smooth interpolation between AMP sequences, and showsthe potential for optimization of peptide'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 1}, page_content='representation of peptides allows forthe generation of novel AMPs in an automated fashion, enablesthe smooth interpolation between AMP sequences, and showsthe potential for optimization of peptide characteristics.Furthermore, we experimentally examine AMPs generated viainterpolation and explore the potential of the method for AMPdesign.2. RESULTS2.1. Model Creation and Generation of ActivePeptides.We implemented and trained a VAE for modelingpeptide sequences with antimicrobial activity. We then usedthe model for generative peptidede novodesign. The generaldesign of the standard VAE architecture is shown inFigure 1.'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 1}, page_content='Figure 1.Machine learning for antimicrobial design. (A) Flow diagram of VAE design. During training, both known and scrambled peptides are fedinto the encoder, generating their latent codes (z). The latent codes are then decoded into the peptide sequence by the decoder. The two terms ofthe loss function (KL and reconstruction loss) and model are updated using stochastic gradient descent. (B) Here, we train the VAE using an AMPdataset, which allows for experimental validation of results using common microbiological assays. From starting input AMP sequences, the encodernetwork converts each peptide into a vector in the latent space, which can be viewed as a continuous AMP representation. Provided a point in thelatent space, the decoder network will output a corresponding AMP sequence. The output sequences can then be experimentally tested for activity.'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 1}, page_content='Figure 2.Antimicrobial activity assays. (A) Three peptides, P1, P2, and P3, were selected from point locations near short active peptides describedin the literature, NA-CATH, KR-12, and Tet110, generating three peptides sequences designated P1, P2, and P3, respectively, and threecorresponding scrambled control peptides (SP1, SP2, and SP3). (B) Experimental assessment of active (P1−P3) AMPs and scrambled (SP1−SP3)pairs onE. coli,A. baumannii, andS. aureus. Peptide concentrations were 0, 0.64, 3.2, 16, 80, and 400μg/mL. Peptide sequences are listed inTable1. Inhibition of colony formation is indicative of AMP activity.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420747'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 2}, page_content='Five-fold cross-validation on diﬀerent network architectureswas tested for best performance as measured by loss, wherenetworks with 512 or 1024 intermediate dimensionsperformed best, with no observable advantage with largernetworks. Training was stopped after 250 epochs as the lossdecreased at a suﬃciently low rate (less than 0.1 per 10epochs;Figure S1) and results did not noticeably improve withlonger periods of training. Theﬁnal state of the model wassaved and used for sampling novel sequences (Figure 2A).Since the model was trained using two sets of sequencescomposed of“active”and“scrambled”peptides, the implicitstarting assumption was that the sequence order, or“peptidegrammar”,20and characteristics dependent on that sequencewere the components that would be learned by the model.Characteristics related to amino acid frequency, such as netcharge, would not directly contribute to model training asthese are also present in the scrambled peptides.Using the generated latent space, new'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 2}, page_content='related to amino acid frequency, such as netcharge, would not directly contribute to model training asthese are also present in the scrambled peptides.Using the generated latent space, new peptides weredecoded using locations near short active peptides describedin the literature. Latent space positions for the peptides NA-CATH (AP00897; its 11-mer N-terminal variant, ATRA-1, hasbeen extensively studied21), KR-12 (AP00608),22and Tet110(AP02874)23were identiﬁed and used to generate threepeptides sequences designated P1, P2, and P3, respectively.From these sequences, three scrambled control peptides (SP1,SP2, and SP3) were produced (Figure 2A), only altering thesequence order while maintaining amino acid frequency andlength (seeTable 1). Each of these peptides was then assessedfor antimicrobial activity againstEscherichia coli,Acinetobacterbaumannii, andStaphylococcus aureus(Figure 2B).By spot-plating the bacteria following incubation with arange of AMP concentrations, half-maximal'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 2}, page_content='activity againstEscherichia coli,Acinetobacterbaumannii, andStaphylococcus aureus(Figure 2B).By spot-plating the bacteria following incubation with arange of AMP concentrations, half-maximal eﬀective concen-tration (EC50) values were calculated from triplicate experi-ments. ForE. coli, the EC50values of P1, P2, and P3 weredetermined to be 2.9, 3.1, and 2.9μg/mL, respectively, whileonly SP3 had a determinable EC50of 3.1μg/mL; the others,SP1 and SP2, did not display signiﬁcant killing at theconcentrations of peptide tested (<400μg/mL). The EC50values of P1, P2, and P3 againstA. baumanniiwere found to be3.1, 5.9, and 1.8μg/mL, respectively. Both SP1 and SP3obtained calculable EC50’s of 10.6 and 19μg/mL, respectively,while SP2 did not display antimicrobial activity at theconcentrations used. For the Gram-positiveS. aureus,theEC50values of P1, P2, and P3 were determined to be 0.4, 6.6,and 2.2μg/mL, respectively. Both SP1 and SP3 obtainedcalculable EC50’s of 10.6 and 16.5μg/mL,'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 2}, page_content='used. For the Gram-positiveS. aureus,theEC50values of P1, P2, and P3 were determined to be 0.4, 6.6,and 2.2μg/mL, respectively. Both SP1 and SP3 obtainedcalculable EC50’s of 10.6 and 16.5μg/mL, respectively, whileTable 1. Peptides Used in the Study and Associated Characteristicsname/series increment sequence length molecular weight net charge hydrophobicity hydrophobic momentP1 RKLKKLWRKFR 11 1558.983 6.997−1.782 0.969P2 RRFVKKVRKLVK 12 1557.008 6.997−0.825 0.976P3 FRWLRKWFRR 10 1550.878 4.998−1.430 1.045SP1 KKRRFRWLKLK 11 1558.983 6.997−1.782 0.161SP2 RKLRKKVFVKRV 12 1557.008 6.997−0.825 0.253SP3 WKRLRWRRFF 10 1550.878 4.998−1.430 0.173series 1 0.00 VIREHKYVLLL 11 1382.712 1.090 0.718 0.366series 1 0.33 LPKIKKTVSTR 11 1270.582 3.997−0.682 0.345series 1 0.66 LLKSGRLLMKI 11 1271.671 2.997 0.736 0.637series 1 0.99 KKIKRFLRKIG 11 1386.793 5.997−0.855 1.015series 2 0.00 GLGIIPHRRYGK 12 1366.632 3.088−0.617 0.325series 2 0.33 GIMSLFKGVLKT 12 1293.631 1.997 0.908 0.589series 2 0.66'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 2}, page_content='2.997 0.736 0.637series 1 0.99 KKIKRFLRKIG 11 1386.793 5.997−0.855 1.015series 2 0.00 GLGIIPHRRYGK 12 1366.632 3.088−0.617 0.325series 2 0.33 GIMSLFKGVLKT 12 1293.631 1.997 0.908 0.589series 2 0.66 GLFKIIKNIFSG 12 1336.640 1.997 0.833 0.676series 2 0.99 KLFRIIKRIFKG 12 1518.955 4.997 0.150 1.075'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 2}, page_content='Figure 3.Latent space plots. Latent space maps were generated from the APD3 database (plus scrambled sequences). Antimicrobial concept vector(black line) with stepwise sampling shown, extending from the scrambled peptide (A) SP1 (green star) to P1 (red star) and (B) SP2 (green star) toP2 (red star). V1 and V2 are latent variables 1 and 2, respectively. Antimicrobial activity (black arrow) is expected to increase or decreaseproportional to the distance from the scrambled/inactive (SP) or active (P) position. Peptide sequences for each point in the space can be found inTable 1.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420748'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 3}, page_content='SP2 was found to be inactive. EC50values are shown inTableS3. Overall, by mean activity, the scrambled peptides wereassessed as signiﬁcantly less active (p< 0.05; Welch’st-test)than those generated by sampling near active peptides in thelatent space. This result is visible in representative platingimages shown inFigure 2B. This result suggested that themodel learned a degree of sequence-dependent informationconsidering that the scrambled peptides maintained identicalnet charge, hydrophobicity, and other sequence order-independent attributes (listed inTable 1). Of note was thesigniﬁcant diﬀerence in hydrophobic moment between P1, P2,and P3 peptides when compared to SP1, SP2, and SP3: meanhydrophobic moment values of 1.0 and 0.2, respectively. Noneof the other biophysical characteristics examined were found tobe signiﬁcantly diﬀerent between the groups.2.2. Interpolation and Dataset Characterization.Totest the eﬀect of interpolation between active and scrambledpeptides, the overall'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 3}, page_content='examined were found tobe signiﬁcantly diﬀerent between the groups.2.2. Interpolation and Dataset Characterization.Totest the eﬀect of interpolation between active and scrambledpeptides, the overall characteristics of the decoded peptideswere examined to determine the eﬀect of diﬀerent samplingpoints in the latent space. A representation of the latent spacegenerated from 2971 active (obtained from APD311) and 2971corresponding scrambled peptide sequences is shown inFigure3A,B. Here, two series of peptides were generated, betweenSP1 and P1 and between SP2 and P2. Linear interpolation wasperformed, and four peptides were generated along what canbe called an antimicrobial concept vector (black line) instepwise sampling (Figure 3). These four peptide sequenceswere generated at increments of 0.33: 0.00, 0.33, 0.66, and0.99. The sequence at 0.00 was generated close to scrambledpeptide (SP) and the sequence at 0.99 was generated close tothe active peptide (P). For each series, peptides were'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 3}, page_content='0.00, 0.33, 0.66, and0.99. The sequence at 0.00 was generated close to scrambledpeptide (SP) and the sequence at 0.99 was generated close tothe active peptide (P). For each series, peptides were chosen atrelative distances of 0.33 and 0.66 along the prediction line orantimicrobial concept vector (Table 1). A series was notgenerated for SP3−P3 since both were found to be active.To determine the diﬀerent distributions of characteristics ofthe peptides generated using interpolation, as shown inFigure3, we performed this for every active and scrambled pair in thedataset (2971 pairings, plus the P1−SP1 and P2−SP2 pairs),resulting in a list of 17,838 sequences each with a location label(scrambled, 0.00, 0.33, 0.66, 0.99, and active; seeTable S2).Attributes that are considered to be important for antimicro-bial activity of AMPs, such as hydrophobic moment,24andothers that are likely uncorrelated with activity, such as length,were calculated as a function of step (location of'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 3}, page_content='be important for antimicro-bial activity of AMPs, such as hydrophobic moment,24andothers that are likely uncorrelated with activity, such as length,were calculated as a function of step (location of sampling).Although the AMP length was not restricted solely to 12 aminoacids, but was trimmed to≤12, the majority of AMPs were atthat length (96%), with the remainder consisting of shorterpeptides (Figure 4A). Lengths were similar, regardless ofsampling location, with each group containing between 95 and98% peptides of length 12.Like length, the amino acid frequency was similar regardlessof sampling location while not being perfectly uniform (Figure4B). This intergroup similarity was also the case for net charge(Figure 4C), amino acid type (Figure S2A), isoelectric point(pI;Figure S2B), and molecular weight (Figure S2C). Withthe exception of net charge, each of these attributes is generallyconsidered to be largely uncorrelated to antimicrobial activityin AMPs. The net charge within the'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 3}, page_content='molecular weight (Figure S2C). Withthe exception of net charge, each of these attributes is generallyconsidered to be largely uncorrelated to antimicrobial activityin AMPs. The net charge within the group was positive (1.24±0.01 (mean±SEM)) ranging between +1.19 and +1.31.While a net positive charge is common among AMPs, between12 and 15% of naturally occurring AMPs are thought to beanionic (net negative charge).25Beyond this, as net charge isan attribute resulting from amino acid frequency and the activeand scrambled group AMPs are identical in this regard, thesameness of the groups in net charge was expected.Peptide hydrophobicity and protein hydrophobicity aredependent on amino acid frequency and structural elements.Within this study and dataset, all groups have a hydrophobicityvalue between 0.08 and 0.2 (overall, 0.13±0.01 (mean±'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 3}, page_content='Figure 4.Comparison of peptide characteristics for the entire generated dataset. (A) Length, (B) amino acid frequency, (C) net charge, (D)hydrophobicity, and (E) hydrophobic moment. For (D) and (E), the middle line is displayed at the median. The sequence list for the generateddataset is provided inTable S2.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420749'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 4}, page_content='SEM)) as measured by the Kyte−Doolittle scale.26Interest-ingly, there was an observable trend upward betweenscrambled and active groups (Figure 4D). When examiningthe larger dataset of active and scrambled peptides, at steps0.00, 0.33, 0.66, and 0.99, the hydrophobicity values weredetermined to be 0.08±0.02, 0.11±0.02, 0.14±0.02, and0.19±0.02, respectively. Both scrambled and active groupswere 0.11±0.02. Since overall hydrophobicity is known to bepositively correlated with antimicrobial activity,24this result isnot surprising; however, interestingly, the trend is notcontinued in scrambled and active groups.To estimate the proportion of the secondary structure of thegroups, the GOR IV method was used, which providedpredicted propensities for helix, beta sheet, and random coil.As expected, the average peptide in the active group ispredicted to be primarily helical, while their scrambledcounterparts have a median predicted helicity of∼15%, withthe steps in between increasing in helicity'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 4}, page_content='the average peptide in the active group ispredicted to be primarily helical, while their scrambledcounterparts have a median predicted helicity of∼15%, withthe steps in between increasing in helicity when samplednearer to active (Figure S3). A related, signiﬁcant trend wasobserved in comparing the predicted hydrophobic momentsbetween the sampling step locations. The hydrophobicmoments were calculated using a default angle of 100°,which assumes a helical secondary structure. Hydrophobicmoments for each group were found to be 0.26±0.002, 0.28±0.002, 0.30±0.003, 0.32±0.003, 0.35±0.003, and 0.38±0.003 for scrambled, 0.00, 0.33, 0.66, 0.99, and active,respectively. This means that the spatial orientations ofhydrophobic and polar amino acids diﬀer between the groupsalong the interpolation path, leading to signiﬁcantly increasedhydrophobic moments (p< 0.01, two-sided Welch’st-test)compared to those of the scrambled peptide set (Figure 4E).Also, importantly, the mean value of each group is'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 4}, page_content='leading to signiﬁcantly increasedhydrophobic moments (p< 0.01, two-sided Welch’st-test)compared to those of the scrambled peptide set (Figure 4E).Also, importantly, the mean value of each group is signiﬁcantlydiﬀerent from each adjacent set, suggesting a gradual shift inspatial orientations along the route from scrambled to active.Next, to quantify the potential antimicrobial activity of thegenerated peptides, we made use of the CAMPR3 AMPprediction tool. We decided to use the CAMPR3 classiﬁer as ithas been empirically determined to be superior to otherclassiﬁers in a recent benchmark report,27best reﬂecting realantimicrobial activity as determined in experiments. Theprobability of antimicrobial activity (P(AMP)) of the AMPswas obtained and plotted inFigure 5A. In a trend similar to thehydrophobic moment described above, theP(AMP) increasedas a function of the sampling step locations between scrambledand active. TheP(AMP) for each group was, as output byCAMPR3, 0.40±0.006, 0.41±0.006,'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 4}, page_content='moment described above, theP(AMP) increasedas a function of the sampling step locations between scrambledand active. TheP(AMP) for each group was, as output byCAMPR3, 0.40±0.006, 0.41±0.006, 0.41±0.006, 0.44±0.006, 0.49±0.007, and 0.50±0.007 for scrambled, 0.00,0.33, 0.66, 0.99, and active, respectively. Meanwhile, the overallP(AMP) was 0.44±0.002 (mean±SEM), suggesting that, onaverage, the peptides generated in this study are predicted tobe inactive (where aP(AMP) of≥0.5 is classiﬁed as active).However, several of these are clearly active, as seen inFigure5B and discussed below. Critically, Welch’st-test rejected thehypothesis that the scrambled and active datasets have equalmeans of activity predictions (p< 0.05), as well as 0.66 and0.99 datasets when compared to the scrambled dataset (p<0.05). Interestingly, at the 0.33 point, the peptides were not ofhigherP(AMP).The relatively lowP(AMP) prediction for those sequencesin the active peptide group may result from the following: (1)the'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 4}, page_content='Interestingly, at the 0.33 point, the peptides were not ofhigherP(AMP).The relatively lowP(AMP) prediction for those sequencesin the active peptide group may result from the following: (1)the sequences in the training set for CAMPR3 are substantiallylonger on average than those used in this study (≤12), (2)short peptides may not contain suﬃcient information for mostmodels to properly assess likelihood of activity, and (3) it ispossible that the truncation of peptides in the APD3-baseddataset did impact real activity. Overall, according to theCAMPR3 predictions, the VAE model can be used tosmoothly interpolate between AMPs predicted to be active(or borderline active) and those that are predicted to beinactive.2.3. Evaluation of Antimicrobial Activity.As describedabove (Section 2.2), following their testing, the peptides P1,P2, SP1, and SP2 were used as markers to test the capability ofinterpolating between the two groups, scrambled and active.Two series of interpolating peptides were'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 4}, page_content='their testing, the peptides P1,P2, SP1, and SP2 were used as markers to test the capability ofinterpolating between the two groups, scrambled and active.Two series of interpolating peptides were generated at fourincrements between each of the scrambled−active pairs of P1−SP1 and P2−SP2 at increments of 0.00, 0.33, 0.66, and 0.99, asshown inFigure 3.T h er e s u l t i n gp e p t i d e sw e r en e x texperimentally investigated by assessing their antimicrobialactivities onE. coli,A. baumannii, andS. aureus(Figure 5B). Byspot-plating the bacteria following incubation with a range ofAMP concentrations, EC50’s were calculated from triplicateexperiments. EC50’s for both series 1 and 2 peptides againsteach of the three species tested are found inTable S4andcorresponding EC50curves are shown inFigure S4.ForE. coli, the EC50values of peptides 0.66 and 0.99 forseries 1 were found to be 308 and 11μg/mL, respectively,while neither preceding peptides in the series displayed'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 4}, page_content='Figure 5.(A) Active AMP prediction using CAMPR3. (B) Peptide series 1 and 2 tested againstE. coli,A. baumannii, andS. aureus.Peptideconcentrations were 0, 0.64, 3.2, 16, 80, and 400μg/mL. Row labels are interpolation step for each peptide series. The absence or inhibition ofcolony formation is indicative of AMP activity.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle\\nhttps://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420750'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='activity. Similarly, for series 2, peptides 0.66 and 0.99 werefound to have EC50’s of 84 and 0.2μg/mL, respectively, whileboth peptides 0.00 and 0.33 were inactive at the concentrationstested. When tested againstA. baumannii, the EC50values ofpeptides 0.66 and 0.99 for series 1 were found to be 400 and36μg/mL, respectively, while neither of the preceding peptides0.00 or 0.33 display activity. For series 2, the EC50’s wereobserved to be 65, 13, and 0.8μg/mL for 0.33, 0.66, and 0.99,respectively; peptide 0.00 was inactive. ForS. aureus,the EC50values of peptides 0.33, 0.66, and 0.99 for series 1 weredetermined as 69, 11, and 0.4μg/mL, respectively, whilepeptide 0.00 did not show activity. For series 2, all of thepeptides were found inactive at the concentrations examined.Overall, these results appear to conﬁrm theP(AMP)predictions by CAMPR3 inFigure 5A, suggesting that thepeptides sampled from regions closer to the active space inlatent space are more active than those decoded in'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='results appear to conﬁrm theP(AMP)predictions by CAMPR3 inFigure 5A, suggesting that thepeptides sampled from regions closer to the active space inlatent space are more active than those decoded in closeproximity to scrambled. This trend is further supported bycomparing the overall mean activities, which show that 0.99group AMPs have a signiﬁcantly lower EC50than those from0.66 (p< 0.05; Welch’st-test). Although the EC50values formany of those in group 0.33 and all of those in group 0.00 arenot available due to insuﬃcient activity for calculation,Figure5B suggests the overall trend of increasing activity along thepath between scrambled and active groups.3. DISCUSSIONThis study demonstrates the use of a VAE forde novopeptidesequence design using AMPs as an example. The VAEdescribed has the potential for sequence design automationprovided that a large database of AMP sequences is available, arequirement similar to other computational techniques. Theincreasing number of both AMP'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='has the potential for sequence design automationprovided that a large database of AMP sequences is available, arequirement similar to other computational techniques. Theincreasing number of both AMP databases and entries perdatabase makes AMPs a viable proof of concept for assessingthe ability of VAEs to generate new peptides. This approach isdistinct from many previously reported design techniques, suchas random mutagenesis, template-based peptide designstrategies, and other random sequence generation methods,where amino acids are drawn from a predeﬁned distributionand then concatenated.5The encoder and decoder of the VAErely on LSTM RNNs that have high-dimensional sequencerepresentations, meaning that it does not simply reproducesequence templates but learns important features present in thetraining dataset.The results achieved suggest that the peptide VAE modelwas capable of generating an internal representation of AMPsfrom the APD3 database, in that the model did not'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='present in thetraining dataset.The results achieved suggest that the peptide VAE modelwas capable of generating an internal representation of AMPsfrom the APD3 database, in that the model did not simplyreproduce the training data, supported by the absence of anypreviously known sequences in those experimentally examined.However, when sampled from similar locations in the latentspace, the decoded peptides do have statistically similardistributions to those found in the training data (either activeor scrambled). Critically, the features generally reported to beimportant for the antimicrobial activity of AMPs, whether forbacterial membrane targeting or otherwise, such as highhydrophobic moment in alpha helical peptides,24are retainedin the peptides decoded near active peptides, while propertiesless correlated with antimicrobial activity, such as length,molecular weight, and pI, are roughly constant regardless oflocation in the latent space. In fact, of all of the peptideattributes'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='correlated with antimicrobial activity, such as length,molecular weight, and pI, are roughly constant regardless oflocation in the latent space. In fact, of all of the peptideattributes examined in the study, the main diﬀerence betweenthe active model-generated sequences and scrambled randomsequences is their amphipathicity, which is illustrated by ahigher hydrophobic moment for the model-generated setsclose to active peptides in the latent space, consistent with theresults reported by Müller et al. utilizing a LSTM RNNmodel.8Since the hydrophobic moment translates back to a regularpattern of charged and hydrophobic residues at speciﬁclocations in the sequence, this demonstrates that this hasbeen learned by the model. It is very likely that the increasinghydrophobic moment yielded higher probabilities in theCAMPR3 AMP prediction model,28suggesting that simplyalternating groups of positively ionizable and hydrophobicamino acids will score highly by the CAMPR3 predictor, ahypothesis'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='probabilities in theCAMPR3 AMP prediction model,28suggesting that simplyalternating groups of positively ionizable and hydrophobicamino acids will score highly by the CAMPR3 predictor, ahypothesis that has been manipulated by Müller et al.,8highlighting the importance of experimentally verifying theantimicrobial activity of generated AMPs. This, along withpreserving the simplicity of the method presented, is a possibleadvantage of avoiding pairing generative models withpredictive models, which has been done elsewhere:29we canavoid any possibility of forcing the generative model to“learn”to tailor its output to what scores highly in a predictive model.To verify antimicrobial activity, the interpolation featureprovided by VAEs was exploited. By interpolating between twoknown peptides, an active peptide found in APD3 and anassumed inactive scrambled partner, increasing gradations ofantimicrobial activity can be sampled at regular steps along theline (Figure 2). This study demonstrated'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='active peptide found in APD3 and anassumed inactive scrambled partner, increasing gradations ofantimicrobial activity can be sampled at regular steps along theline (Figure 2). This study demonstrated that two series ofpeptides, between P1−SP1 and P2−SP2, sampled at four stepseach could yield an increasing spectrum of activity betweentwo points. The results indicate that not only did the VAEactually learn to recognize the grammar of amphipathicity inpeptides but also it may have learned the underlying grammarof what features contribute to antimicrobial activity. Thecriticalﬁnding is that the VAE learned the grammar ofamphipathicity and antimicrobial activity in peptides, whichcan then be tuned in a controllable manner.As observed in these studies, the intermediates of P3−SP3displayed high antimicrobial activity as both the predicted andscrambled peptides. With the P1, P2, SP1, and SP2 constructs,the reduced activity of the scrambled peptides can likely beattributed to a change in the'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='antimicrobial activity as both the predicted andscrambled peptides. With the P1, P2, SP1, and SP2 constructs,the reduced activity of the scrambled peptides can likely beattributed to a change in the hydrophobic moment and,therefore, the ability for the peptide to readily insert into thebacterial membrane. For the P3−SP3 pair, however, thesigniﬁcant change in hydrophobic moment does not appear toaﬀect the antimicrobial activity. The P3−SP3 peptides arecomposed of multiple tryptophan and arginine residues. In apublication by Wessolowski et al., a similar phenotype wasobserved with short peptides of similar composition,suggesting antimicrobial activity through noncanonical mech-anisms.30While their consistent activity as native andscrambled peptides is interesting, neighboring peptides werenot examined in these studies as similar activity was expected.This observation, however, illustrates the potential need forpairing the VAE with aﬁltering algorithm to remove certainclasses of AMPs'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='examined in these studies as similar activity was expected.This observation, however, illustrates the potential need forpairing the VAE with aﬁltering algorithm to remove certainclasses of AMPs depending on the application.As conﬁrmation of the active AMP prediction model,28thetwo interpolated series of AMPs were tested againstE. coli,A.baumannii, andS. aureusand, overall, yielded the same trendidentiﬁed by the AMP prediction, where peptides sampleddisplayed more activity as a function of their proximity toknown active peptides. Interestingly, the peptide series mayshow a Gram bias, in that the series 2 peptides did not displayany measurable activity againstS. aureusbut were active againstGram-negative bacteriaE. coliandA. baumannii.MoreACS Omegahttp://pubs.acs.org/journal/acsodfArticle'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 5}, page_content='https://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420751'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='investigation is required to determine if this result waspredictable prior to synthesis. Although, for simplicity, nopredictive model was paired with VAE in the current study,various other groups using generative models for chemical orpeptide generation have output shuttled through predictivemodels prior to use.18Such a pairing could be developed forfuture applications to potentially wield some control overGram bias in antimicrobial activity and other attributes ofpeptides prior to experiment assessment.4. CONCLUSIONSThe results of the study demonstrate the use of a VAE for theautomated generation of novel AMPs. The model was trainedon thousands of known and scrambled AMP sequences fromAPD3 and created a readily usable tool, the form of acontinuous representation of peptides. The latent spaceallowed us to generate novel AMPs in a controllable fashion,which we experimentally veriﬁed to be active. Despite theshortcomings of this proof-of-concept study, including thefocus on truncated'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='spaceallowed us to generate novel AMPs in a controllable fashion,which we experimentally veriﬁed to be active. Despite theshortcomings of this proof-of-concept study, including thefocus on truncated peptides, this favorable result will drivefuture work in the application of VAEs toward various peptideand protein-designing projects, including full-length AMPs.This work suggests that the VAE could operate with greaterutility when paired with a predictive model, or othercharacteristic classiﬁer, in a manner described by Gómez-Bombarelli et al. for molecule generation to better predictAMP activity and for optimization purposes. Beyond this, it isunclear whether the VAE technique is speciﬁcally well suitedfor peptide and protein applications or could other recentlydeveloped and improved generative deep learning techniquessuch as GANs result in superior performance. This idea, use ofa conditional VAE or otherwise pairing with a characteristicclassiﬁer, and assessment of VAE-generated AMPs'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='deep learning techniquessuch as GANs result in superior performance. This idea, use ofa conditional VAE or otherwise pairing with a characteristicclassiﬁer, and assessment of VAE-generated AMPs on otherbacteria will be the subject of future research.5. EXPERIMENTAL SECTION5.1. Dataset and Preprocessing.AMP sequences andother information, including activity (Gram-negative bacteria,Gram-positive bacteria, cancer cells, etc.), were downloadedfrom APD3,11yielding 2971 peptide sequences (accessed inmid-2018). Despite theﬂaws of APD3, such as highhomogeneity and incomplete annotation of post-translationallymodiﬁed sequences, it remains to be one of largest AMPdatabases available. Since synthesis of long peptides can beprohibitively expensive, and to reduce the computational loaddownstream, the sequences were truncated such that allpeptides were≤12 amino acids in length, retaining the N-terminal sequence. The length of≤12 amino acids was chosenbased on several reports that have shown that'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='sequences were truncated such that allpeptides were≤12 amino acids in length, retaining the N-terminal sequence. The length of≤12 amino acids was chosenbased on several reports that have shown that short (10−12amino acids), truncated AMPs are capable of maintaining highlevels of antimicrobial activity.4,31,32The list of trimmedpeptide sequences was then duplicated and randomlyscrambled with the stringi R package21function stri_rand_-shuﬄe to create a scrambled peptide pair for each peptidefound in APD3, resulting in a list of 5942 sequences. Duplicateactive peptides were retained as all scrambled pairings wereunique. The sequences were then tokenized (each residueseparated; seeTable S1) and represented by a one-hotencoding scheme using binary vectors with length equal to thesize of the amino acid vocabulary:“<end>”,“a”,“c”,“d”,“e”,“f’”,“g”,“h”,“i”,“k”,“l”,“m”,“n”,“p”,“q”,“r”,“s”,“t”,“v”,“w”,“x”, and“y”, where“x”is an unknown amino acid present insome of the APD3 sequences. This'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='the amino acid vocabulary:“<end>”,“a”,“c”,“d”,“e”,“f’”,“g”,“h”,“i”,“k”,“l”,“m”,“n”,“p”,“q”,“r”,“s”,“t”,“v”,“w”,“x”, and“y”, where“x”is an unknown amino acid present insome of the APD3 sequences. This resulted in a 3D datamatrix of dimensions 5942, 24, and 12 for the number ofsequences, length of the vocabulary, and feature vector length,respectively.5.2. Training and Sequence Generation.The archi-tecture of the VAE was implemented as described by Bowmanet al.,33including use of character dropout, which replaces arandom fraction of characters (listed inSection 5.1) with an<unk> character and utilization of Kullback−Leibler (KL)annealing, comprising an additional variable weight added tothe cost function KL term while training. The loss function, asdescribed by Bowman et al., was composed of reconstructionloss and KL loss to penalize poor reconstruction of the data bythe decoder and encoder output representations of z (latentspace variables) that are diﬀerent from a standard'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='composed of reconstructionloss and KL loss to penalize poor reconstruction of the data bythe decoder and encoder output representations of z (latentspace variables) that are diﬀerent from a standard normaldistribution, respectively (Figure 1A). Training stoppagecriteria were met when loss values did not decrease for 10iterations through the entire dataset (epochs). The prepro-cessed data were encoded into vectors using LSTMs. Theencoder LSTM was paired with a decoder LSTM to dosequence-to-sequence learning. The decoder results wereconverted from binary one-hot encoded vectors to peptidesequences (Figure 1B). The VAE was trained using the Keras34library with a TensorFlow35backend via the Adam optimizer.The number of neurons for the LSTM layers found in bothencoder and decoder was both set to 1024. All models weretrained on an Ubuntu workstation with a Nvidia GeforceGTX1070 GPU. The LSTM RNNs used in the decoder (andencoder) are stochastic, meaning that decoding from the samepoint in'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='to 1024. All models weretrained on an Ubuntu workstation with a Nvidia GeforceGTX1070 GPU. The LSTM RNNs used in the decoder (andencoder) are stochastic, meaning that decoding from the samepoint in the latent space may result in a diﬀerent peptide beinggenerated and is dependent on the random seed set prior torunning. Sequence sampling was performed using linearinterpolation between active peptides and their scrambledpairs, which is expected to be inactive or less functional as anAMP against target bacteria. For each pairing, four points weregenerated at equally spaced increments of 0.33: 0.00, 0.33,0.66, and 0.99, where the scrambled pair is encoded at point 0and the active pair is encoded at 1 such that points 0.00 and0.99 are adjacent to AMPs in the dataset, while points decodedat 0.33 and 0.66 are intermediates. Here, it was theorized thatantimicrobial activity would correspond to the latent spaceposition or distance from either the active (position 1.0) orinactive/scrambled'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='0.33 and 0.66 are intermediates. Here, it was theorized thatantimicrobial activity would correspond to the latent spaceposition or distance from either the active (position 1.0) orinactive/scrambled peptide (position 0.0).5.3. Examination of Sequence Characteristics.Thecharacteristics of sequences generated, including peptidelength, amino acid composition, net charge, hydrophobicity,and hydrophobic moment, were assessed using the Peptides Rpackage36and Python library modlAMP.37Prediction of thethree-state protein secondary structure (helix, beta, and coil)was carried out using the Garnier−Osguthorpe−Robson(GOR) IV method38implemented in the Decipher Rpackage.39The probability of AMP activity (P(AMP)) waspredicted using CAMPR327by uploading sequences to itsserver. Exported results from CAMPR3, as well as character-istic values output from the Peptides package and modlAMP,were visualized using ggplot2 in the tidyverse R package.405.4. Antimicrobial Assays.Peptides synthesized for use'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='as well as character-istic values output from the Peptides package and modlAMP,were visualized using ggplot2 in the tidyverse R package.405.4. Antimicrobial Assays.Peptides synthesized for use inthis study are listed inTable 1. Peptides were synthesized byGenscript, Inc. (Piscataway, NJ), and each was conﬁrmed tohave greater than 90% purity. Lyophilized peptides weresolubilized in water, aliquoted, and stored at−20°C.Overnight cultures ofE. coliBL21,A. baumanniiATCC17978, andS. aureusATCC 12600 were grown in LysogenyACS Omegahttp://pubs.acs.org/journal/acsodfArticle'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 6}, page_content='https://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420752'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 7}, page_content='Broth (LB). EC50assays were performed as previouslydescribed.41Overnight cultures of bacteria were enumeratedby a standard curve using optical density vs CFU/mL. Cultureswere resuspended in 10 mM sodium phosphate buﬀer (pH 7)to aﬁnal concentration of∼2×106CFU/mL. Fordetermination of EC50values, peptides were assessed at arange of concentrations (0.64−400μg/mL) with bacteria for 1h at 37°C. Following incubation, bacteria were spotted forcolony counting as previously described;42full spread platingwas performed for conﬁrmation of colony counting.■ASSOCIATED CONTENT*sıSupporting InformationThe Supporting Information is available free of charge athttps://pubs.acs.org/doi/10.1021/acsomega.0c00442.EC50values for P1, P2, and P3 and their scrambled pairs(Table S3); EC50values corresponding to the survivalcurves found in Figure S3 (Table S4); representative losscurve from the training of the VAE (Figure S1);comparison of peptide characteristics (Figure S2);secondary structure prediction of'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 7}, page_content='the survivalcurves found in Figure S3 (Table S4); representative losscurve from the training of the VAE (Figure S1);comparison of peptide characteristics (Figure S2);secondary structure prediction of generated peptidesacross each group (Figure S3); and EC50curves for thepeptide series examined (Figure S4) (PDF)Input dataset for training (Table S1) (XLSX)Generated peptides (Table S2) (XLSX)■AUTHOR INFORMATIONCorresponding AuthorScott A. Walper−Center for Bio/Molecular Science&Engineering (Code 6900), U.S. Naval Research Laboratory,Washington, D.C. 20375, United States;orcid.org/0000-0002-9436-3456; Email:scott.walper@nrl.navy.milAuthorScott N. Dean−National Research Council Associate,Washington, D.C. 20001, United StatesComplete contact information is available at:https://pubs.acs.org/10.1021/acsomega.0c00442Author ContributionsThe manuscript was written through contributions of allauthors. All authors have given approval to theﬁnal version ofthe manuscript.NotesThe authors declare no'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 7}, page_content='ContributionsThe manuscript was written through contributions of allauthors. All authors have given approval to theﬁnal version ofthe manuscript.NotesThe authors declare no competingﬁnancial interest.■ACKNOWLEDGMENTSS.N.D. acknowledges the funding support through theNational Research Council Research Associateship Programand additional funding through the Jerome and Isabella KarleDistinguished Scholar Fellowship provided by the NavalResearch Laboratory. S.A.W. was supported by Core fundsprovided to the Naval Research Laboratory through Oﬃce ofNaval Research (WU#MA041−06−41).■ABBREVIATIONSAMPs, antimicrobial peptides; VAE, variational autoencoder;RNN, recursive neural network; LSTM, long short-termmemory; EC50, half-maximal eﬀective concentration; SMILES,simpliﬁed molecular-input line-entry system■REFERENCES(1) O’Neill, J. Antimicrobial Resistance: Tackling a crisis for thehealth and wealth of nations.Rev. Antimicrob. Resist.2014, 1.(2) Lewies, A.; Du Plessis, L. H.; Wentzel, J. F.'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 7}, page_content='system■REFERENCES(1) O’Neill, J. Antimicrobial Resistance: Tackling a crisis for thehealth and wealth of nations.Rev. Antimicrob. Resist.2014, 1.(2) Lewies, A.; Du Plessis, L. H.; Wentzel, J. F. AntimicrobialPeptides: the Achilles’Heel of Antibiotic Resistance?ProbioticsAntimicrob. Proteins2019, 370−381.(3) Chung, E. M. C.; Dean, S. N.; Propst, C. N.; Bishop, B. M.; vanHoek, M. L. Komodo dragon-inspired synthetic peptide DRGN-1promotes wound-healing of a mixed-biofilm infected wound.npjBiofilms Microbiomes2017,3, 9.(4) Dean, S. N.; Bishop, B. M.; van Hoek, M. L. Natural andsynthetic cathelicidin peptides with anti-microbial and anti-biofilmactivity against Staphylococcus aureus.BMC Microbiol.2011,11, 114.(5) Torres, M. D. T.; Sothiselvam, S.; Lu, T. K.; de la Fuente-Nunez,C. Peptide Design Principles for Antimicrobial Applications.J. Mol.Biol.2019, 3547.(6) Cardoso, M. H.; Oshiro, K. G. N.; Rezende, S. B.; Cândido, E.S.; Franco, O. L. The Structure/Function Relationship in'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 7}, page_content='Design Principles for Antimicrobial Applications.J. Mol.Biol.2019, 3547.(6) Cardoso, M. H.; Oshiro, K. G. N.; Rezende, S. B.; Cândido, E.S.; Franco, O. L. The Structure/Function Relationship in Anti-microbial Peptides: What Can we Obtain From Structural Data?Adv.Protein Chem. Struct. Biol.2018,112, 359−384.(7) Lima, A. N.; Philot, E. A.; Trossini, G. H. G.; Scott, L. P. B.;Maltarollo, V. G.; Honorio, K. M. Use of machine learning approachesfor novel drug discovery.Expert Opin. Drug Discovery2016,11, 225−239.(8) Müller, A. T.; Hiss, J. A.; Schneider, G. Recurrent NeuralNetwork Model for Constructive Peptide Design.J. Chem. Inf. Model.2018,58, 472−479.(9) Nagarajan, D.; Nagarajan, T.; Roy, N.; Kulkarni, O.;Ravichandran, S.; Mishra, M.; Chakravortty, D.; Chandra, N.Computational antimicrobial peptide design and evaluation againstmultidrug-resistant clinical isolates of bacteria.J. Biol. Chem.2018,293, 3492−3509.(10) Schneider, P.; Müller, A. T.; Gabernet, G.; Button, A. L.;Posselt,'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 7}, page_content='peptide design and evaluation againstmultidrug-resistant clinical isolates of bacteria.J. Biol. Chem.2018,293, 3492−3509.(10) Schneider, P.; Müller, A. T.; Gabernet, G.; Button, A. L.;Posselt, G.; Wessler, S.; Hiss, J. A.; Schneider, G. Hybrid NetworkModel for″Deep Learning″of Chemical Data: Application toAntimicrobial Peptides.Mol. Inf.2017,36, 1600011.(11) Wang, G.; Li, X.; Wang, Z. APD3: the antimicrobial peptidedatabase as a tool for research and education.Nucleic Acids Res.2016,44, D1087−D1093.(12) Lee, H.-T.; Lee, C.-C.; Yang, J.-R.; Lai, J. Z. C.; Chang, K. Y. Alarge-scale structural classification of antimicrobial peptides.BioMedRes. Int.2015,2015,1−6.(13) Porto, W. F.; Pires, A. S.; Franco, O. L. Computational tools forexploring sequence databases as a resource for antimicrobial peptides.Biotechnol. Adv.2017,35, 337−349.(14) Nagarajan, D.; Nagarajan, T.; Nanajkar, N.; Chandra, N. Auniform in vitro efficacy dataset to guide antimicrobial peptide design.Data2019,4, 27.(15)'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 7}, page_content='peptides.Biotechnol. Adv.2017,35, 337−349.(14) Nagarajan, D.; Nagarajan, T.; Nanajkar, N.; Chandra, N. Auniform in vitro efficacy dataset to guide antimicrobial peptide design.Data2019,4, 27.(15) Gers, F. A.; Schmidhuber, E. LSTM recurrent networks learnsimple context-free and context-sensitive languages.IEEE Trans.Neural Networks2001,12, 1333−1340.(16) Kingma, D. P.; Welling, M. Auto-encoding variational bayes.2013, arXiv:1312.6114. arXiv.org e-Print archive.https://arxiv.org/abs/1312.6114.(17) White, T. Sampling generative networks.2016,arXiv:1609.04468. arXiv.org e-Print archive.https://arxiv.org/abs/1609.04468.(18) Gömez-Bombarelli, R.; Wei, J. N.; Duvenaud, D.; Hernández-Lobato, J. M.; Sánchez-Lengeling, B.; Sheberla, D.; Aguilera-Iparraguirre, J.; Hirzel, T. D.; Adams, R. P.; Aspuru-Guzik, A.Automatic chemical design using a data-driven continuous repre-sentation of molecules.ACS Cent. Sci.2018,4, 268−276.(19) Kusner, M. J.; Paige, B.; Hernández-Lobato, J. M.'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 7}, page_content='R. P.; Aspuru-Guzik, A.Automatic chemical design using a data-driven continuous repre-sentation of molecules.ACS Cent. Sci.2018,4, 268−276.(19) Kusner, M. J.; Paige, B.; Hernández-Lobato, J. M. Grammarvariational autoencoder.InProceedings of the 34th InternationalConference on Machine Learning-Volume 70, JMLR.org: 2017; pp1945−1954.(20) Loose, C.; Jensen, K.; Rigoutsos, I.; Stephanopoulos, G. Alinguistic model for the rational design of antimicrobial peptides.Nature2006,443, 867.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 7}, page_content='https://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420753'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 8}, page_content='(21) Dean, S. N.; Bishop, B. M.; van Hoek, M. L. Susceptibility ofPseudomonas aeruginosa Biofilm to Alpha-Helical Peptides: D-enantiomer of LL-37.Front. Microbiol.2011,2, 128.(22) Wang, G. Structures of human host defense cathelicidin LL-37and its smallest antimicrobial peptide KR-12 in lipid micelles.J. Biol.Chem.2008,283, 32637−32643.(23) Hilpert, K.; Elliott, M.; Jenssen, H.; Kindrachuk, J.; Fjell, C. D.;Körner, J.; Winkler, D. F. H.; Weaver, L. L.; Henklein, P.; Ulrich, A.S.; Chiang, S. H. Y.; Farmer, S. W.; Pante, N.; Volkmer, R.; Hancock,R. E. W. Screening and characterization of surface-tethered cationicpeptides for antimicrobial activity.Chem. Biol.2009,16, 58−69.(24) Pathak, N.; Salas-Auvert, R.; Ruche, G.; Janna, M. h.;McCarthy, D.; Harrison, R. G. Comparison of the effects ofhydrophobicity, amphiphilicity, andα-helicity on the activities ofantimicrobial peptides.Proteins: Struct., Funct., Bioinf.1995,22, 182−186.(25) Harris, F.; Dennison, S. R.; Phoenix, D. A.'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 8}, page_content='effects ofhydrophobicity, amphiphilicity, andα-helicity on the activities ofantimicrobial peptides.Proteins: Struct., Funct., Bioinf.1995,22, 182−186.(25) Harris, F.; Dennison, S. R.; Phoenix, D. A. Anionicantimicrobial peptides from eukaryotic organisms.Curr. ProteinPept. Sci.2009,10, 585−606.(26) Kyte, J.; Doolittle, R. F. A simple method for displaying thehydropathic character of a protein.J. Mol. Biol.1982,157, 105−132.(27) Gabere, M. N.; Noble, W. S. Empirical comparison of web-based antimicrobial peptide prediction tools.Bioinformatics2017,33,1921−1929.(28) Waghu, F. H.; Barai, R. S.; Gurung, P.; Idicula-Thomas, S.CAMPR3: a database on sequences, structures and signatures ofantimicrobial peptides.Nucleic Acids Res.2016,44, D1094−D1097.(29) Xu, Y.; Lin, K.; Wang, S.; Wang, L.; Cai, C.; Song, C.; Lai, L.;Pei, J. Deep learning for molecular generation.Future Med. Chem.2019, 567.(30) Wessolowski, A.; Bienert, M.; Dathe, M. Antimicrobial activityof arginine- and tryptophan-rich'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 8}, page_content='C.; Song, C.; Lai, L.;Pei, J. Deep learning for molecular generation.Future Med. Chem.2019, 567.(30) Wessolowski, A.; Bienert, M.; Dathe, M. Antimicrobial activityof arginine- and tryptophan-rich hexapeptides: the effects of aromaticclusters, D-amino acid substitution and cyclization.J. Pept. Res.2004,64, 159−169.(31) Bruni, N.; Capucchio, M.; Biasibetti, E.; Pessione, E.;Cirrincione, S.; Giraudo, L.; Corona, A.; Dosio, F. Antimicrobialactivity of lactoferrin-related peptides and applications in human andveterinary medicine.Molecules2016,21, 752.(32) Kasetty, G.; Papareddy, P.; Kalle, M.; Rydengård, V.; Mörgelin,M.; Albiger, B.; Malmsten, M.; Schmidtchen, A. Structure-activitystudies and therapeutic potential of host defense peptides of humanthrombin.Antimicrob. Agents Chemother.2011,55, 2880−2890.(33) Bowman, S. R.; Vilnis, L.; Vinyals, O.; Dai, A. M.; Jozefowicz,R.; Bengio, S. Generating sentences from a continuous space.2015,arXiv:1511.06349. arXiv.org e-Print'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 8}, page_content='Chemother.2011,55, 2880−2890.(33) Bowman, S. R.; Vilnis, L.; Vinyals, O.; Dai, A. M.; Jozefowicz,R.; Bengio, S. Generating sentences from a continuous space.2015,arXiv:1511.06349. arXiv.org e-Print archive.https://arxiv.org/abs/1511.06349.(34) Chollet, F.Keras.https://github.com/fchollet/keras.(35) Abadi, M.; Barham, P.; Chen, J.; Chen, Z.; Davis, A.; Dean, J.;Devin, M.; Ghemawat, S.; Irving, G.; Isard, M.Tensorﬂow: A systemfor large-scale machine learning.In12th {USENIX} Symposium onOperating Systems Design and Implementation ({OSDI} 16); USENIXAssociation: 2016; pp 265−283.(36) Osorio, D.; Rondón-Villarrea, P.; Torres, R. Peptides: a packagefor data mining of antimicrobial peptides.R Journal2015,7,4−14.(37) Müller, A. T.; Gabernet, G.; Hiss, J. A.; Schneider, G.modlAMP: Python for antimicrobial peptides.Bioinformatics2017,33, 2753−2755.(38) Garnier, J.; Gibrat, J.-F.; Robson, B. GOR method forpredicting protein secondary structure from amino acid sequence.InMethods in enzymology,'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 8}, page_content='peptides.Bioinformatics2017,33, 2753−2755.(38) Garnier, J.; Gibrat, J.-F.; Robson, B. GOR method forpredicting protein secondary structure from amino acid sequence.InMethods in enzymology, Elsevier: 1996; Vol.266, pp 540−553,DOI: 10.1016/S0076-6879(96)66034-0.(39) Wright, E. S. DECIPHER: harnessing local sequence context toimprove protein multiple sequence alignment.BMC Bioinf.2015,16,322.(40) Wickham, H.Tidyverse: Easily install and load’tidyverse’packages.R package version; R Core Team: 2017,1(1).(41) Chung, M. C.; Dean, S. N.; van Hoek, M. L. Acyl carrierprotein is a bacterial cytoplasmic target of cationic antimicrobialpeptide LL-37.Biochem. J.2015,470, 243−253.(42) Sieuwerts, S.; de Bok, F. A. M.; Mols, E.; de vos, W. M.; vanHylckama Vlieg, J. E. T. A simple and fast method for determiningcolony forming units.Lett. Appl. Microbiol.2008,47, 275−278.ACS Omegahttp://pubs.acs.org/journal/acsodfArticle'),\n",
       " Document(metadata={'source': 'Vae_paper.pdf', 'page': 8}, page_content='https://dx.doi.org/10.1021/acsomega.0c00442ACS Omega2020, 5, 20746−2075420754')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
    "text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000,chunk_overlap=200)\n",
    "documents = text_splitter.split_documents(docs)\n",
    "documents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Error raised by inference API HTTP code: 404, {\"error\":\"model \\\"llama2\\\" not found, try pulling it first\"}",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[7], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain_community\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01membeddings\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m OllamaEmbeddings\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain_community\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mvectorstores\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m FAISS\n\u001b[1;32m----> 3\u001b[0m db\u001b[38;5;241m=\u001b[39m\u001b[43mFAISS\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_documents\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdocuments\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mOllamaEmbeddings\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Shreshth\\rag_project\\rag\\.venv\\lib\\site-packages\\langchain_core\\vectorstores\\base.py:830\u001b[0m, in \u001b[0;36mVectorStore.from_documents\u001b[1;34m(cls, documents, embedding, **kwargs)\u001b[0m\n\u001b[0;32m    828\u001b[0m texts \u001b[38;5;241m=\u001b[39m [d\u001b[38;5;241m.\u001b[39mpage_content \u001b[38;5;28;01mfor\u001b[39;00m d \u001b[38;5;129;01min\u001b[39;00m documents]\n\u001b[0;32m    829\u001b[0m metadatas \u001b[38;5;241m=\u001b[39m [d\u001b[38;5;241m.\u001b[39mmetadata \u001b[38;5;28;01mfor\u001b[39;00m d \u001b[38;5;129;01min\u001b[39;00m documents]\n\u001b[1;32m--> 830\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39mfrom_texts(texts, embedding, metadatas\u001b[38;5;241m=\u001b[39mmetadatas, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Shreshth\\rag_project\\rag\\.venv\\lib\\site-packages\\langchain_community\\vectorstores\\faiss.py:1041\u001b[0m, in \u001b[0;36mFAISS.from_texts\u001b[1;34m(cls, texts, embedding, metadatas, ids, **kwargs)\u001b[0m\n\u001b[0;32m   1014\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[0;32m   1015\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mfrom_texts\u001b[39m(\n\u001b[0;32m   1016\u001b[0m     \u001b[38;5;28mcls\u001b[39m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1021\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[0;32m   1022\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m FAISS:\n\u001b[0;32m   1023\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Construct FAISS wrapper from raw documents.\u001b[39;00m\n\u001b[0;32m   1024\u001b[0m \n\u001b[0;32m   1025\u001b[0m \u001b[38;5;124;03m    This is a user friendly interface that:\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1039\u001b[0m \u001b[38;5;124;03m            faiss = FAISS.from_texts(texts, embeddings)\u001b[39;00m\n\u001b[0;32m   1040\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m-> 1041\u001b[0m     embeddings \u001b[38;5;241m=\u001b[39m \u001b[43membedding\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43membed_documents\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtexts\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1042\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m__from(\n\u001b[0;32m   1043\u001b[0m         texts,\n\u001b[0;32m   1044\u001b[0m         embeddings,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1048\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[0;32m   1049\u001b[0m     )\n",
      "File \u001b[1;32mc:\\Shreshth\\rag_project\\rag\\.venv\\lib\\site-packages\\langchain_community\\embeddings\\ollama.py:208\u001b[0m, in \u001b[0;36mOllamaEmbeddings.embed_documents\u001b[1;34m(self, texts)\u001b[0m\n\u001b[0;32m    199\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Embed documents using an Ollama deployed embedding model.\u001b[39;00m\n\u001b[0;32m    200\u001b[0m \n\u001b[0;32m    201\u001b[0m \u001b[38;5;124;03mArgs:\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    205\u001b[0m \u001b[38;5;124;03m    List of embeddings, one for each text.\u001b[39;00m\n\u001b[0;32m    206\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    207\u001b[0m instruction_pairs \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39membed_instruction\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00mtext\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m text \u001b[38;5;129;01min\u001b[39;00m texts]\n\u001b[1;32m--> 208\u001b[0m embeddings \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_embed\u001b[49m\u001b[43m(\u001b[49m\u001b[43minstruction_pairs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    209\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m embeddings\n",
      "File \u001b[1;32mc:\\Shreshth\\rag_project\\rag\\.venv\\lib\\site-packages\\langchain_community\\embeddings\\ollama.py:196\u001b[0m, in \u001b[0;36mOllamaEmbeddings._embed\u001b[1;34m(self, input)\u001b[0m\n\u001b[0;32m    194\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    195\u001b[0m     iter_ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28minput\u001b[39m\n\u001b[1;32m--> 196\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m [\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_process_emb_response(prompt) \u001b[38;5;28;01mfor\u001b[39;00m prompt \u001b[38;5;129;01min\u001b[39;00m iter_]\n",
      "File \u001b[1;32mc:\\Shreshth\\rag_project\\rag\\.venv\\lib\\site-packages\\langchain_community\\embeddings\\ollama.py:196\u001b[0m, in \u001b[0;36m<listcomp>\u001b[1;34m(.0)\u001b[0m\n\u001b[0;32m    194\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    195\u001b[0m     iter_ \u001b[38;5;241m=\u001b[39m \u001b[38;5;28minput\u001b[39m\n\u001b[1;32m--> 196\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m [\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_process_emb_response\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprompt\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m prompt \u001b[38;5;129;01min\u001b[39;00m iter_]\n",
      "File \u001b[1;32mc:\\Shreshth\\rag_project\\rag\\.venv\\lib\\site-packages\\langchain_community\\embeddings\\ollama.py:170\u001b[0m, in \u001b[0;36mOllamaEmbeddings._process_emb_response\u001b[1;34m(self, input)\u001b[0m\n\u001b[0;32m    167\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError raised by inference endpoint: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    169\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m res\u001b[38;5;241m.\u001b[39mstatus_code \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m200\u001b[39m:\n\u001b[1;32m--> 170\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[0;32m    171\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mError raised by inference API HTTP code: \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m, \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m    172\u001b[0m         \u001b[38;5;241m%\u001b[39m (res\u001b[38;5;241m.\u001b[39mstatus_code, res\u001b[38;5;241m.\u001b[39mtext)\n\u001b[0;32m    173\u001b[0m     )\n\u001b[0;32m    174\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    175\u001b[0m     t \u001b[38;5;241m=\u001b[39m res\u001b[38;5;241m.\u001b[39mjson()\n",
      "\u001b[1;31mValueError\u001b[0m: Error raised by inference API HTTP code: 404, {\"error\":\"model \\\"llama2\\\" not found, try pulling it first\"}"
     ]
    }
   ],
   "source": [
    "from langchain_community.embeddings import OllamaEmbeddings\n",
    "from langchain_community.vectorstores import FAISS\n",
    "db=FAISS.from_documents(documents, OllamaEmbeddings())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query=\"Who are the authors of this research paper?\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
